{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3db10756",
   "metadata": {},
   "source": [
    "Code to reproduce plots from: https://www.sciencedirect.com/science/article/pii/S1465324921007453 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fd703efe",
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = '/Users/admin/HLA/RUNS_haplomat/1835/frequencies.dat'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8f0677f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def haplomat_output_parce(hfs_dat, n):\n",
    "    \n",
    "    with open(hfs_dat) as f:\n",
    "        cretan_haplotypes = [x.replace('\\n', '').replace('g', '').split() for x in f.readlines()]\n",
    "\n",
    "    cretan_haplotypes = [( '-'.join(sorted(x[0].split('~'))  ), float(x[1])) for x in cretan_haplotypes]\n",
    "\n",
    "    cretan_haplotypes = {x[0]:x[1] for x in cretan_haplotypes}\n",
    "    cretan_haplotypes = {k:v for k,v in cretan_haplotypes.items() if v >= 1/(2*n)}\n",
    "    \n",
    "    return cretan_haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b534a6da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d904a5d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### After we run the Hapl-o-Mat with the g grouping without missing data we calculate the allele frequencies\n",
    "\n",
    "def af_estimation_from_HaploMat_g_group(hfs_dat, hla_gene, num_of_people, alleles=None, counts=None):\n",
    "    \n",
    "    '''\n",
    "    !!!README!!!\n",
    "    \n",
    "    This function calculates the allele frequency of the HLA genes A, B, C, DRB1, DQB1 and DPB1 \n",
    "    if the haplotypes have them. \n",
    "    \n",
    "    Hapl-o-Mat orders the genes in the haplotypes alphabetically.\n",
    "    \n",
    "    The haplotypes with 4 genes should have the HLA genes below in this order only:\n",
    "    A~B~C~DRB1\n",
    "    \n",
    "    The haplotypes with 5 genes should have the HLA genes below in this order only:\n",
    "    A~B~C~DQB1~DRB1\n",
    "    \n",
    "    The haplotypes with 6 genes should have the HLA genes below in this order only:\n",
    "    A~B~C~DPB1~DQB1~DRB1\n",
    "    \n",
    "    Other combinations of the HLA genes can not be used as input.    \n",
    "    \n",
    "    This function returns the estimated allele frequencies/counts and alleles of the sample from Hapl-o-Mat.\n",
    "    '''\n",
    "    \n",
    "    if hla_gene not in ('A', 'B', 'C', 'DRB1', 'DQB1', 'DPB1'):\n",
    "        raise Exception(\"The hla_gene can only be 'A', 'B', 'C', 'DRB1', 'DQB1' or 'DPB1'\")\n",
    "        \n",
    "    if counts and alleles:\n",
    "        raise Exception('Only 1 output can come from this function.')\n",
    "    \n",
    "    with open(hfs_dat, 'r') as f:\n",
    "        f = f.readlines()\n",
    "        \n",
    "    f = [x.strip('\\n').split('\\t') for x in f]\n",
    "    allele_freq = [(x[0].split('~'), float(x[1])) for x in f]\n",
    "    allele_freq = [([y.strip('g') for y in x[0]],x[1]) for x in allele_freq]\n",
    "    \n",
    "    if hla_gene == 'A':\n",
    "        n = 0        \n",
    "    elif hla_gene == 'B':\n",
    "        n = 1        \n",
    "    elif hla_gene == 'C':\n",
    "        n = 2        \n",
    "    \n",
    "    elif hla_gene == 'DRB1': # Hapl-o-Mat orders the genes in the haplotypes alphabetically.\n",
    "        if len(allele_freq[0][0]) == 4:     \n",
    "            n = 3\n",
    "        elif len(allele_freq[0][0]) == 5:     \n",
    "            n = 4    \n",
    "        elif len(allele_freq[0][0]) == 6:\n",
    "            n = 5\n",
    "        else:\n",
    "            raise Exception('The haplotypes should have 4, 5 or 6 genes in a specific order. Read the documentation.')\n",
    "    \n",
    "    elif hla_gene == 'DQB1':\n",
    "        if len(allele_freq[0][0]) == 4:     \n",
    "            raise Exception('DQB1 should not be in a haplotype with only 4 genes. Read the documentation.')\n",
    "        elif len(allele_freq[0][0]) == 5:     \n",
    "            n = 3    \n",
    "        elif len(allele_freq[0][0]) == 6:\n",
    "            n = 4\n",
    "        else:\n",
    "            raise Exception('The haplotypes should have 4, 5 or 6 genes in a specific order. Read the documentation.')                 \n",
    "    \n",
    "    elif hla_gene == 'DPB1':\n",
    "        if len(allele_freq[0][0]) == 4:     \n",
    "            raise Exception('DPB1 should not be in a haplotype with only 4 genes. Read the documentation.')\n",
    "        elif len(allele_freq[0][0]) == 5:     \n",
    "            raise Exception('DPB1 should not be in a haplotype with only 5 genes. Read the documentation.')    \n",
    "        elif len(allele_freq[0][0]) == 6:\n",
    "            n = 3\n",
    "        else:\n",
    "            raise Exception('The haplotypes should have 4, 5 or 6 genes in a specific order. Read the documentation.')\n",
    "    \n",
    "    HLA_gene = {el:0 for el in [x[0][n] for x in allele_freq]}\n",
    "    \n",
    "    for x in allele_freq:\n",
    "        HLA_gene[x[0][n]] += x[1]\n",
    "    \n",
    "    HLA_gene = {k:v for k, v in sorted(HLA_gene.items(), key=lambda item: item[1], reverse=True)}\n",
    "    \n",
    "    if alleles:\n",
    "        return list(HLA_gene.keys())\n",
    "    \n",
    "    elif counts:        \n",
    "        return {k:round(v*num_of_people) for k,v in HLA_gene.items()} \n",
    "        # No need to multiply by 2, because of the 2 genotypes on MAC input\n",
    "    \n",
    "    else:\n",
    "        return HLA_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2d680d26",
   "metadata": {},
   "outputs": [],
   "source": [
    "def chi2(gene_final_cretan_study, gene_final_freqnet, common_alleles, people_in_study, population, sig_afs=None, counts=None):\n",
    "    \n",
    "    if not len(gene_final_cretan_study) == len(gene_final_freqnet) == len(common_alleles):\n",
    "        raise Exception('Not the same length of common genes')\n",
    "    \n",
    "    gene = common_alleles[0].split('*')[0]\n",
    "    \n",
    "    cretan_counts = [x*people_in_study*2 for x in gene_final_cretan_study]\n",
    "    cretan_counts = [np.round(x) for x in cretan_counts]\n",
    "\n",
    "    af_pop = population['pop'].to_list()[0]\n",
    "    freqnet_num = population['sample_size'].to_list()[0]   \n",
    "    \n",
    "    pop_counts = [x*freqnet_num*2 for x in gene_final_freqnet]\n",
    "    \n",
    "    comparison_list = [list(a) for a in zip(cretan_counts, pop_counts)]\n",
    "    \n",
    "    if counts:   \n",
    "        obs = []\n",
    "        for x in range(len(comparison_list)):\n",
    "            obs.append(np.round(np.array([comparison_list[x], [people_in_study*2-comparison_list[x][0],freqnet_num*2-comparison_list[x][1]]])))\n",
    "        \n",
    "        obs = [np.flip(obs[x], 0).T for x in range(len(obs))]\n",
    "        \n",
    "        return obs\n",
    "    \n",
    "    comparison_list_af = [list(a) for a in zip(gene_final_cretan_study, gene_final_freqnet)]\n",
    "        \n",
    "    pvalues = []\n",
    "\n",
    "    for x in range(len(comparison_list)):\n",
    "        obs = np.round(np.array([comparison_list[x], [people_in_study*2-comparison_list[x][0],freqnet_num*2-comparison_list[x][1]]]))\n",
    "        chi2, p, df, exp = chi2_contingency(obs)\n",
    "        pvalues.append(p)\n",
    "\n",
    "    pval_dict = dict(zip(common_alleles, pvalues))\n",
    "\n",
    "    wanted_alleles = {k:v for k,v in pval_dict.items() if v<0.05}\n",
    "\n",
    "    initial_dict = dict(zip(common_alleles, zip(gene_final_cretan_study, gene_final_freqnet)))\n",
    "    significant_dict = {k:v for z,w in wanted_alleles.items() for k,v in initial_dict.items() if z in k}\n",
    "\n",
    "    if sig_afs:            \n",
    "        significant_dict = {k:v for k,v in significant_dict.items() if v[0] > round(1/(2*people_in_study), 6) and v[1] > round(1/(2*freqnet_num), 4)}\n",
    "\n",
    "    common_alleles_sig = list(significant_dict.keys())\n",
    "    gene_final_cretan_study_sig = [x[0] for x in significant_dict.values()]\n",
    "    gene_final_freqnet_sig = [x[1] for x in significant_dict.values()]\n",
    "\n",
    "    return gene_final_cretan_study_sig, gene_final_freqnet_sig, common_alleles_sig "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "205be61c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_common_g_alleles(cretans, population, HLA_gene):\n",
    "    \n",
    "    alleles_allelefreq_net = population[population['gene']==HLA_gene]['allele'].tolist()  \n",
    "    common_alleles = list(set(alleles_allelefreq_net).intersection(list(cretans.keys())))\n",
    "    \n",
    "    alleles_freqnet = dict(zip(population[population['gene']==HLA_gene].sort_values('af', ascending=False)['allele'].tolist(), population[population['gene']==HLA_gene].sort_values('af', ascending=False)['af'].tolist()))\n",
    "\n",
    "    alleles_final_Cretans = []\n",
    "\n",
    "    for x in common_alleles:\n",
    "        alleles_final_Cretans.append(cretans[x])\n",
    "\n",
    "    tmp_hla_alleles_Cretans = dict(zip(common_alleles, alleles_final_Cretans))\n",
    "    tmp_hla_alleles_Cretans = {k: v for k, v in sorted(tmp_hla_alleles_Cretans.items(), key=lambda item: item[1], reverse=True)}\n",
    "\n",
    "    tmp_alleles_freqnet = {}\n",
    "\n",
    "    for x in tmp_hla_alleles_Cretans.keys():\n",
    "        tmp_alleles_freqnet[x] = alleles_freqnet[x]\n",
    "\n",
    "    common_alleles = list(tmp_hla_alleles_Cretans.keys())\n",
    "    alleles_final_freqnet = list(tmp_alleles_freqnet.values())\n",
    "    alleles_final_Cretans = list(tmp_hla_alleles_Cretans.values())\n",
    "    \n",
    "    return alleles_final_Cretans, alleles_final_freqnet, common_alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "701b0417",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = haplomat_output_parce(fn, 1835)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6bc6023c",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = af_estimation_from_HaploMat_g_group(fn, 'A', 1835)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d61a81e7",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'g4_path' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [19], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m A_g_grouped_1835_Cretans \u001b[38;5;241m=\u001b[39m af_estimation_from_HaploMat_g_group(\u001b[43mg4_path\u001b[49m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;241m1835\u001b[39m)\n\u001b[1;32m      2\u001b[0m B_g_grouped_1835_Cretans \u001b[38;5;241m=\u001b[39m af_estimation_from_HaploMat_g_group(g4_path, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mB\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;241m1835\u001b[39m)\n\u001b[1;32m      3\u001b[0m C_g_grouped_1835_Cretans \u001b[38;5;241m=\u001b[39m af_estimation_from_HaploMat_g_group(g4_path, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mC\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;241m1835\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'g4_path' is not defined"
     ]
    }
   ],
   "source": [
    "A_g_grouped_1835_Cretans = af_estimation_from_HaploMat_g_group(g4_path, 'A', 1835)\n",
    "B_g_grouped_1835_Cretans = af_estimation_from_HaploMat_g_group(g4_path, 'B', 1835)\n",
    "C_g_grouped_1835_Cretans = af_estimation_from_HaploMat_g_group(g4_path, 'C', 1835)\n",
    "DRB1_g_grouped_1835_Cretans = af_estimation_from_HaploMat_g_group(g4_path, 'DRB1', 1835)\n",
    "DQB1_g_grouped_1835_Cretans = af_estimation_from_HaploMat_g_group(g5_path, 'DQB1', 1835)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8ea918cc",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'A_g_grouped_1835_Cretans' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [18], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# ### Let's compare the HLA allelic frequencies of all minorities in Germany with the 1835 Cretans ###\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m A_final_1835_DKMSALL_GER, A_final_freqnet_1835_DKMSALL_GER, common_A_1835_DKMSALL_GER \u001b[38;5;241m=\u001b[39m get_common_g_alleles(\u001b[43mA_g_grouped_1835_Cretans\u001b[49m, DKMS_all_final, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      4\u001b[0m B_final_1835_DKMSALL_GER, B_final_freqnet_1835_DKMSALL_GER, common_B_1835_DKMSALL_GER \u001b[38;5;241m=\u001b[39m get_common_g_alleles(B_g_grouped_1835_Cretans, DKMS_all_final, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mB\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      5\u001b[0m C_final_1835_DKMSALL_GER, C_final_freqnet_1835_DKMSALL_GER, common_C_1835_DKMSALL_GER \u001b[38;5;241m=\u001b[39m get_common_g_alleles(C_g_grouped_1835_Cretans, DKMS_all_final, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mC\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'A_g_grouped_1835_Cretans' is not defined"
     ]
    }
   ],
   "source": [
    "# ### Let's compare the HLA allelic frequencies of all minorities in Germany with the 1835 Cretans ###\n",
    "\n",
    "A_final_1835_DKMSALL_GER, A_final_freqnet_1835_DKMSALL_GER, common_A_1835_DKMSALL_GER = get_common_g_alleles(A_g_grouped_1835_Cretans, DKMS_all_final, 'A')\n",
    "B_final_1835_DKMSALL_GER, B_final_freqnet_1835_DKMSALL_GER, common_B_1835_DKMSALL_GER = get_common_g_alleles(B_g_grouped_1835_Cretans, DKMS_all_final, 'B')\n",
    "C_final_1835_DKMSALL_GER, C_final_freqnet_1835_DKMSALL_GER, common_C_1835_DKMSALL_GER = get_common_g_alleles(C_g_grouped_1835_Cretans, DKMS_all_final, 'C')\n",
    "DRB1_final_1835_DKMSALL_GER, DRB1_final_freqnet_1835_DKMSALL_GER, common_DRB1_1835_DKMSALL_GER = get_common_g_alleles(DRB1_g_grouped_1835_Cretans, DKMS_all_final, 'DRB1')\n",
    "\n",
    "# ### Let's compare the HLA allelic frequencies of all minorities without the Greeks in Germany with the 1835 Cretans ###\n",
    "\n",
    "A_final_1835_DKMSNOGR_GER, A_final_freqnet_1835_DKMSNOGR_GER, common_A_1835_DKMSNOGR_GER = get_common_g_alleles(A_g_grouped_1835_Cretans, DKMS_final, 'A')\n",
    "B_final_1835_DKMSNOGR_GER, B_final_freqnet_1835_DKMSNOGR_GER, common_B_1835_DKMSNOGR_GER = get_common_g_alleles(B_g_grouped_1835_Cretans, DKMS_final, 'B')\n",
    "C_final_1835_DKMSNOGR_GER, C_final_freqnet_1835_DKMSNOGR_GER, common_C_1835_DKMSNOGR_GER = get_common_g_alleles(C_g_grouped_1835_Cretans, DKMS_final, 'C')\n",
    "DRB1_final_1835_DKMSNOGR_GER, DRB1_final_freqnet_1835_DKMSNOGR_GER, common_DRB1_1835_DKMSNOGR_GER = get_common_g_alleles(DRB1_g_grouped_1835_Cretans, DKMS_final, 'DRB1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87316711",
   "metadata": {},
   "outputs": [],
   "source": [
    "e,f,g=chi2(A_final_1835_DKMSALL_GER, A_final_freqnet_1835_DKMSALL_GER, common_A_1835_DKMSALL_GER, 1835, DKMS_all_final, sig_afs=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "db62c2b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_comparison_diagram(gene_final_cretan_study, gene_final_freqnet, common_alleles, people_in_study, population, sig_af=None, sig_p=None):\n",
    "\n",
    "    if not len(gene_final_cretan_study) == len(gene_final_freqnet) == len(common_alleles):\n",
    "        raise Exception('Not the same length of common genes')\n",
    "    \n",
    "    af_pop = population['pop'].to_list()[0]\n",
    "    freqnet_num = population['sample_size'].to_list()[0]\n",
    "    \n",
    "    if sig_af and not sig_p:\n",
    "        tmp = dict(zip(common_alleles, zip(gene_final_cretan_study, gene_final_freqnet)))\n",
    "        tmp = {k:v for k,v in tmp.items() if v[0] > round(1/(2*people_in_study), 6) and v[1] > round(1/(2*freqnet_num), 4)}\n",
    "        \n",
    "        gene_final_cretan_study = [v[0] for k,v in tmp.items()]\n",
    "        gene_final_freqnet = [v[1] for k,v in tmp.items()]\n",
    "        common_alleles = list(tmp.keys()) \n",
    "        \n",
    "    common_alleles_num = len(common_alleles)\n",
    "    gene = common_alleles[0].split('*')[0]\n",
    "    index = np.arange(common_alleles_num) \n",
    "    bar_width = 0.35\n",
    "    \n",
    "    fig, ax = plt.subplots()\n",
    "\n",
    "    if people_in_study == 94:\n",
    "        Cretans_94 = ax.bar(index, gene_final_cretan_study, bar_width, label=\"94 Cretans\", color = 'orange')\n",
    "\n",
    "    if 675 < people_in_study < 690:    \n",
    "        Cretans_689 = ax.bar(index, gene_final_cretan_study, bar_width, label=f\"{people_in_study} Cretans\", color = 'green')\n",
    "     \n",
    "    if 1100 < people_in_study < 1215:    \n",
    "        Cretans_1204 = ax.bar(index, gene_final_cretan_study, bar_width, label=f\"{people_in_study} Cretans\", color = 'red')\n",
    "    \n",
    "    if 1700 < people_in_study < 2001:    \n",
    "#         Cretans_1835 = ax.bar(index, gene_final_cretan_study, bar_width, label=f\"{people_in_study} Cretans\", color = 'goldenrod')\n",
    "        Cretans_1835 = ax.bar(index, gene_final_cretan_study, bar_width, label=f\"{people_in_study} Cretans\", color = 'dimgrey')\n",
    "    \n",
    "#     freq_net = ax.bar(index+bar_width, gene_final_freqnet, bar_width, label=f\"{freqnet_num} {af_pop}\", color = 'blue')\n",
    "    freq_net = ax.bar(index+bar_width, gene_final_freqnet, bar_width, label=f\"{freqnet_num} {af_pop}\", color = 'darkgrey')\n",
    "\n",
    "    ax.set_ylabel('Allele Frequency', fontsize=13)\n",
    "    \n",
    "#     if sig_p and not sig_af:\n",
    "#         ax.set_title(f'Comparison of HLA-{gene} frequencies with p < 0.05', fontsize=15)\n",
    "#     elif sig_af and not sig_p:\n",
    "#         ax.set_title(f'Comparison of HLA-{gene} frequencies with af > 1/2n', fontsize=15)\n",
    "#     elif sig_af and sig_p:\n",
    "#         ax.set_title(f'Comparison of HLA-{gene} frequencies with p < 0.05 and af > 1/2n', fontsize=11)\n",
    "#     else:\n",
    "#         ax.set_title(f'Comparison of HLA-{gene} frequencies', fontsize=15)\n",
    "        \n",
    "    ax.set_xticks(index + bar_width / 2)\n",
    "    \n",
    "    if gene == 'B':\n",
    "        font = 7\n",
    "    else:\n",
    "        font = 11\n",
    "        \n",
    "    ax.set_xticklabels(common_alleles, rotation=90, fontsize=font)\n",
    "    ax.legend()\n",
    "    \n",
    "#     fig.tight_layout()    \n",
    "#     fig.set_dpi(300)\n",
    "    \n",
    "#     if sig_p and not sig_af:\n",
    "#         if af_pop == 'DKMS with Greeks' or af_pop == 'DKMS without Greeks':      \n",
    "#             plt.savefig(f'{af_pop}_comparison_{people_in_study}_freqnet/HLA-{gene}/Comparison_of_significant_HLA-{gene}_frequencies_{people_in_study}_{af_pop}.png')\n",
    "#         else:\n",
    "#             plt.savefig(f\"{af_pop.split(' ')[3]}_in_Germany_comparison_{people_in_study}_freqnet/HLA-{gene}/Comparison_of_significant_HLA-{gene}_frequencies_{people_in_study}_{af_pop}.png\")\n",
    "#     elif sig_af and not sig_p:\n",
    "#         if af_pop == 'DKMS with Greeks' or af_pop == 'DKMS without Greeks':      \n",
    "#             plt.savefig(f'{af_pop}_comparison_{people_in_study}_freqnet/HLA-{gene}/Comparison_of_HLA-{gene}_frequencies_{people_in_study}_{af_pop}_with_af_greater_than_1_to_2n.png')\n",
    "#         else:\n",
    "#             plt.savefig(f\"{af_pop.split(' ')[3]}_in_Germany_comparison_{people_in_study}_freqnet/HLA-{gene}/Comparison_of_HLA-{gene}_frequencies_{people_in_study}_{af_pop}_with_af_greater_than_1_to_2n.png\")\n",
    "#     elif sig_af and sig_p:\n",
    "#         if af_pop == 'DKMS with Greeks' or af_pop == 'DKMS without Greeks':      \n",
    "#             plt.savefig(f'{af_pop}_comparison_{people_in_study}_freqnet/HLA-{gene}/Comparison_of_significant_HLA-{gene}_frequencies_{people_in_study}_{af_pop}_with_af_greater_than_1_to_2n.png')\n",
    "#         else:\n",
    "#             plt.savefig(f\"{af_pop.split(' ')[3]}_in_Germany_comparison_{people_in_study}_freqnet/HLA-{gene}/Comparison_of_significant_HLA-{gene}_frequencies_{people_in_study}_{af_pop}_with_af_greater_than_1_to_2n.png\")\n",
    "#     else:\n",
    "#         if af_pop == 'DKMS with Greeks' or af_pop == 'DKMS without Greeks':      \n",
    "#             plt.savefig(f'{af_pop}_comparison_{people_in_study}_freqnet/HLA-{gene}/Comparison_of_HLA-{gene}_frequencies_{people_in_study}_{af_pop}.png')\n",
    "#         else:\n",
    "#             plt.savefig(f\"{af_pop.split(' ')[3]}_in_Germany_comparison_{people_in_study}_freqnet/HLA-{gene}/Comparison_of_HLA-{gene}_frequencies_{people_in_study}_{af_pop}.png\")\n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc1266b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_comparison_diagram(e,f,g,1835, DKMS_all_final, sig_p=True, sig_af=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57bf5f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########################################################################################################################\n",
    "###########################################################################################################################\n",
    "###########################################################################################################################\n",
    "###########################################################################################################################\n",
    "###########################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "632205b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter, defaultdict\n",
    "import pandas as pd\n",
    "import itertools\n",
    "from itertools import chain\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "fbd2ce2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "###############################################################\n",
    "##      Calculates the frequency of each allele              ##\n",
    "###############################################################\n",
    "def allele_freq(pop_txt, #txt file=> columns = [A,A,,B,B...], rows = [01:01,02:01,....]\n",
    "                classII=None, #classII = [DRB1,DQB1,DPB1]\n",
    "               ):\n",
    "    \n",
    "    '''\n",
    "    The Cretan  HLA profile should include all the HLA genes (A,B,C,DRB1,DQB1,DPB1). \n",
    "    So we keep all the individuals (1835) who are fully genotyped on these 6 HLA genes.\n",
    "    '''\n",
    "    \n",
    "    with open('/Users/vasou/Downloads/'+pop_txt, 'r') as f:\n",
    "        af = f.readlines()\n",
    "        \n",
    "    results = defaultdict()\n",
    "    af = [x.strip('\\n').split('\\t') for x in af]\n",
    "    num_people=len(af)\n",
    "    \n",
    "    alleles = {'A':[1,2], #allele:indexes on x(for x in af)\n",
    "               'B':[3,4],\n",
    "               'C':[5,6],\n",
    "               'DRB1':[7,8],\n",
    "               'DQB1':[9,10],\n",
    "               'DPB1':[11,12],\n",
    "              }\n",
    "\n",
    "    for key in alleles:\n",
    "        results[f'N_{key}'] = dict(Counter([x[alleles[key][0]] for x in af[1:]] + [x[alleles[key][1]] for x in af[1:]]))\n",
    "        results[f'N_{key}'] = {k:v for k,v in sorted(results[f'N_{key}'].items(), key=lambda item:item[1], reverse = True)}\n",
    "        results[f'AF_{key}'] = {k:v/(2*num_people) for k,v in results[f'N_{key}'].items()}\n",
    "        results[f'{key}'] = [f'{key}*'+x for x in list(results[f'N_{key}'].keys())]\n",
    "    \n",
    "    if not classII:   \n",
    "        keys = ['A', 'N_A', 'AF_A', 'B', 'N_B', 'AF_B', 'C', 'N_C', 'AF_C']\n",
    "        nest = [results[key].values() if '_' in key else results[key] for key in keys]\n",
    "        target_class = pd.DataFrame((_ for _ in itertools.zip_longest(*nest,fillvalue = '0')), columns=keys)\n",
    "        target_class[['N_A','N_B','N_C']] =  target_class[['N_A','N_B','N_C']].astype(int)\n",
    "        return target_class\n",
    "    else:\n",
    "        keys = ['DRB1', 'N_DRB1', 'AF_DRB1', 'DQB1', 'N_DQB1', 'AF_DQB1', 'DPB1', 'N_DPB1', 'AF_DPB1']\n",
    "        nest = [results[key].values() if '_' in key else results[key] for key in keys]\n",
    "        target_class = pd.DataFrame((_ for _ in itertools.zip_longest(*nest,fillvalue = '0')), columns=keys)\n",
    "        target_class[['N_DRB1','N_DQB1','N_DPB1']] =  target_class[['N_DRB1','N_DQB1','N_DPB1']].astype(int)\n",
    "        return target_class\n",
    "\n",
    "\n",
    "classI = allele_freq('pop.txt', classII=False).head(10)\n",
    "classII = allele_freq('pop.txt', classII=True).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "fcfd6f05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of unique samples in Crete_428CBUs_120723.xlsx: 384\n",
      "The number of unique samples in Crete_428CBUs_120723.xlsx: 384\n",
      "The number of unique samples in Crete_428CBUs_120723.xlsx: 384\n"
     ]
    }
   ],
   "source": [
    "####################################################################################\n",
    "##       Create dict: key = ID, value = dict(nomos:chania, A:01:01,B....)         ##\n",
    "####################################################################################\n",
    "def get_cretan_population(excel_file,#data with necessary columns: ID, REGIONAL UNIT OF DESCENT 2ND GENERATION_75%, alleles\n",
    "                          sheet_name = None,#read sheets\n",
    "                          Crete_75=None, #take the samples that the 75% belongs to crete \n",
    "                          Crete_50_and_75=None,\n",
    "                          municipality_75=None,\n",
    "                          municipality_50_and_75=None):\n",
    "    \n",
    "    if sum(x for x in [Crete_75, Crete_50_and_75, municipality_75, municipality_50_and_75] if x) > 1:\n",
    "        raise Exception('Only 1 value must be True.')\n",
    "\n",
    "    if sheet_name:\n",
    "        df = pd.read_excel(excel_file, sheet_name=sheet_name)\n",
    "    else:\n",
    "        df = pd.read_excel(excel_file)\n",
    "    df.fillna('', inplace=True)  \n",
    "    print(f'The number of unique samples in {excel_file}: {len(pd.unique(df[\"ID\"]))}')\n",
    "\n",
    "    targets = ['REGIONAL UNIT OF DESCENT 2ND GENERATION_75%',\\\n",
    "             'A','B','C',\\\n",
    "             'DRB1','DQB1','DPB1']\n",
    "\n",
    "    data = defaultdict(lambda:defaultdict())\n",
    "    \n",
    "    for row in df.iterrows():\n",
    "        for target in targets:\n",
    "            if target in ['REGIONAL UNIT OF DESCENT 2ND GENERATION_75%']:\n",
    "                data[row[1]['ID']][target] = row[1]['REGIONAL UNIT OF DESCENT 2ND GENERATION_75%']\n",
    "            else:\n",
    "                data[row[1]['ID']][target] = (row[1][target+'1'], row[1][target+'2']) \\\n",
    "                if 'D' not in target else (row[1][target+'_1'], row[1][target+'_2'])\n",
    "                \n",
    "    if Crete_50_and_75:\n",
    "        return data\n",
    "    \n",
    "    elif Crete_75:\n",
    "        return {k:v for k,v in data.items() if not any(\n",
    "                x in v['REGIONAL UNIT OF DESCENT 2ND GENERATION_75%']  \n",
    "                for x in ['','not found','ABROAD']\n",
    "                )\n",
    "               }\n",
    "    \n",
    "    elif municipality_50_and_75: #pali ta abroad vgazei?\n",
    "        return {k:v for k,v in data.items() if v['REGIONAL UNIT OF DESCENT 2ND GENERATION_75%'] not in (\n",
    "           '50% ΕΚΤΟΣ ΚΡΗΤΗΣ/25% ΗΡΑΚΛΕΙΟΥ & 25% ΡΕΘΥΜΝΟΥ',\n",
    "           '50% ΕΚΤΟΣ ΚΡΗΤΗΣ/25% ΛΑΣΙΘΙΟΥ & 25% ΗΡΑΚΛΕΙΟΥ',\n",
    "           '50% ΕΚΤΟΣ ΚΡΗΤΗΣ/25% ΡΕΘΥΜΝΟΥ & 25% ΗΡΑΚΛΕΙΟΥ',\n",
    "           '50% ΕΚΤΟΣ ΚΡΗΤΗΣ/25% ΡΕΘΥΜΝΟΥ & 25% ΛΑΣΙΘΙΟΥ',\n",
    "           '50% ΕΚΤΟΣ ΚΡΗΤΗΣ/25% ΧΑΝΙΩΝ & 25% ΗΡΑΚΛΕΙΟΥ',\n",
    "           '50% ΕΚΤΟΣ ΚΡΗΤΗΣ/25% ΧΑΝΙΩΝ & 25% ΛΑΣΙΘΙΟΥ', \n",
    "           '50% ΕΚΤΟΣ ΚΡΗΤΗΣ/25% ΧΑΝΙΩΝ & 25% ΡΕΘΥΜΝΟΥ',\n",
    "           '25% ΧΑΝΙΩΝ & 25% ΛΑΣΙΘΙΟΥ/50% ΕΚΤΟΣ ΚΡΗΤΗΣ',\n",
    "           '25% ΡΕΘΥΜΝΟΥ & 25% ΗΡΑΚΛΕΙΟΥ/50% ΕΚΤΟΣ ΚΡΗΤΗΣ',\n",
    "           '50% ΕΚΤΟΣ ΚΡΗΤΗΣ/25% ΗΡΑΚΛΕΙΟΥ & 25% ΑΝΩΓΕΙΩΝ',\n",
    "           '25% ΡΕΘΥΜΝΟΥ & 25% ΛΑΣΙΘΙΟΥ/50% ΕΚΤΟΣ ΚΡΗΤΗΣ')}\n",
    "    \n",
    "    elif municipality_75:                \n",
    "        return {k:v for k,v in data.items() if v['REGIONAL UNIT OF DESCENT 2ND GENERATION_75%'] in ('ΗΡΑΚΛΕΙΟΥ', 'ΛΑΣΙΘΙΟΥ', 'ΡΕΘΥΜΝΟΥ', 'ΧΑΝΙΩΝ')}\n",
    "\n",
    "Crete_50_and_75 = get_cretan_population('Crete_428CBUs_120723.xlsx', Crete_50_and_75=True)\n",
    "Crete_75 = get_cretan_population('Crete_428CBUs_120723.xlsx', Crete_75=True)\n",
    "#municipality_50_and_75 = get_cretan_population('final samples_prefecture analyses_111120.xlsx',  '1926_excl_relations', municipality_50_and_75=True)\n",
    "municipality_75 = get_cretan_population('Crete_428CBUs_120723.xlsx', municipality_75=True)#n_Cretans = len(Crete_75)\n",
    "#n_Prefectures = len(municipality_75)\n",
    "\n",
    "#Chania_75 = {k:v for k,v in municipality_75.items() if v['ΝΟΜΟΣ_ΚΑΤΑΓΩΓΗΣ_min75%']=='ΧΑΝΙΩΝ'}\n",
    "#Rethymno_75 = {k:v for k,v in municipality_75.items() if v['ΝΟΜΟΣ_ΚΑΤΑΓΩΓΗΣ_min75%']=='ΡΕΘΥΜΝΟΥ'}\n",
    "#Heraklion_75 = {k:v for k,v in municipality_75.items() if v['ΝΟΜΟΣ_ΚΑΤΑΓΩΓΗΣ_min75%']=='ΗΡΑΚΛΕΙΟΥ'}\n",
    "Lasithi_75 = {k:v for k,v in municipality_75.items() if v['REGIONAL UNIT OF DESCENT 2ND GENERATION_75%']=='ΛΑΣΙΘΙΟΥ'}\n",
    "#n_Chania = len(Chania_75)\n",
    "#n_Rethymno = len(Rethymno_75)\n",
    "#n_Heraklion = len(Heraklion_75)\n",
    "#n_Lasithi = len(Lasithi_75)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "ae7a4900",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'': [4, 1],\n",
       " '01:01': [0, 2],\n",
       " '02:01': [0, 2],\n",
       " '02:05': [0, 1],\n",
       " '11:01': [1, 1],\n",
       " '23:01': [0, 1],\n",
       " '30:01': [0, 1],\n",
       " '32:01': [0, 1]}"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###########################################################################################\n",
    "##    Return a dict with key:allelle or nan and value:[homozygosity, heterozygosity]     ##\n",
    "###########################################################################################\n",
    "def find_allele_homozygosity(pop_dict, #dict key = ID, value = dict(nomos:chania, A:01:01,B....)\n",
    "                             hla_gene # A,B,C....\n",
    "                            ):\n",
    "    \n",
    "    if hla_gene not in ('A', 'B', 'C', 'DRB1', 'DQB1', 'DPB1'):\n",
    "        raise Exception('The HLA gene can only be A, B, C, DRB1, DQB1 or DPB1.')\n",
    "    \n",
    "    n_Cretans = len(pop_dict)\n",
    "    alleles = sorted(list(set(chain(*[v[hla_gene] for k,v in pop_dict.items()]))))\n",
    "    allele_homozygosity = {x:0 for x in alleles}\n",
    "    allele_heterozygosity = {x:0 for x in alleles}\n",
    "    homozygosity_check = [v[hla_gene] for k,v in pop_dict.items()]\n",
    "    \n",
    "    for x in alleles:\n",
    "        for y in homozygosity_check:\n",
    "            if y[0]==y[1]==x:\n",
    "                allele_homozygosity[x] += 1\n",
    "            if y[0]==x and y[0]!=y[1] or y[1]==x and y[0]!=y[1]:\n",
    "                allele_heterozygosity[x] += 1\n",
    "                \n",
    "    result = {}\n",
    "    \n",
    "    for key in (allele_homozygosity.keys() | allele_heterozygosity.keys()):\n",
    "        if key in allele_homozygosity: result.setdefault(key, []).append(allele_homozygosity[key])\n",
    "        if key in allele_heterozygosity: result.setdefault(key, []).append(allele_heterozygosity[key])\n",
    "    result = {k:v for k,v in sorted(result.items(), key = lambda item:item[0])}\n",
    "    \n",
    "    return result\n",
    "\n",
    "find_allele_homozygosity(Lasithi_75,#created by get_cretan_population()\n",
    "                         'A'\n",
    "                        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "665a30d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Homozygotes on 10 Cretans: 6 (60.0%)\n",
      "Homozygotes on A: 5 (50.0%)\n",
      "Homozygotes on B: 5 (50.0%)\n",
      "Homozygotes on C: 5 (50.0%)\n",
      "Homozygotes on DRB1: 5 (50.0%)\n",
      "Homozygotes on DQB1: 4 (40.0%)\n",
      "Homozygotes on DPB1: 3 (30.0%)\n"
     ]
    }
   ],
   "source": [
    "###########################################################################################\n",
    "##    Return a dict with key:allelle or nan and value:[homozygosity, heterozygosity]     ##\n",
    "###########################################################################################\n",
    "def get_homozygosity(pop_dict, #dict key = ID, value = dict(nomos:chania, A:01:01,B....)\n",
    "                     get_ids=None#=1, if you want to take the IDs of homozygotes\n",
    "                    ):\n",
    "\n",
    "    homozygotes_ids = []\n",
    "    homozygotes = defaultdict(lambda:0)\n",
    "    genes = ['A', 'B', 'C', 'DRB1', 'DQB1', 'DPB1']\n",
    "    n_Cretans = len(pop_dict)\n",
    "\n",
    "    for x in list(pop_dict.keys()):\n",
    "        for gene in genes:\n",
    "            \n",
    "            if pop_dict[x][gene][0] == pop_dict[x][gene][1]:\n",
    "                homozygotes[gene] += 1\n",
    "                homozygotes_ids.append(x)\n",
    "    if get_ids:\n",
    "        return {x:pop_dict[x] for x in list(set(homozygotes_ids))}\n",
    "    \n",
    "    print(f'Total Homozygotes on {n_Cretans} Cretans:', len(set(homozygotes_ids)), f'({round((len(set(homozygotes_ids))/n_Cretans)*100,2)}%)')\n",
    "    for gene in genes:\n",
    "        print(f'Homozygotes on {gene}: {homozygotes[gene]} ({round((homozygotes[gene]/n_Cretans)*100,2)}%)')\n",
    "\n",
    "\n",
    "get_homozygosity(Lasithi_75, get_ids=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa81cc5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################################################################################################\n",
    "## den pianw akoma:                                                                                                    ##\n",
    "## create_pop_excel => download excels                                                                                 ##\n",
    "## find_high_homozygosity => ??????????????                                                                            ##\n",
    "## preprocess_af => This function returns the allelic frequencies of the populations of allelefrequencies.net          ##\n",
    "## get_diagram_af_net => This function creates the plots for the allele frequencies of the                             ##\n",
    "##                       DKMS minorities as a whole and 1 by 1.                                                        ##\n",
    "##                                                                                                                     ##\n",
    "#########################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "601d07f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### After we run the Hapl-o-Mat with the g grouping without missing data we calculate the allele frequencies\n",
    "\n",
    "def af_estimation_from_HaploMat_g_group(hfs_dat, hla_gene, num_of_people, alleles=None, counts=None):\n",
    "    \n",
    "    '''\n",
    "    !!!README!!!\n",
    "    \n",
    "    This function calculates the allele frequency of the HLA genes A, B, C, DRB1, DQB1 and DPB1 \n",
    "    if the haplotypes have them. \n",
    "    \n",
    "    Hapl-o-Mat orders the genes in the haplotypes alphabetically.\n",
    "    \n",
    "    The haplotypes with 4 genes should have the HLA genes below in this order only:\n",
    "    A~B~C~DRB1\n",
    "    \n",
    "    The haplotypes with 5 genes should have the HLA genes below in this order only:\n",
    "    A~B~C~DQB1~DRB1\n",
    "    \n",
    "    The haplotypes with 6 genes should have the HLA genes below in this order only:\n",
    "    A~B~C~DPB1~DQB1~DRB1\n",
    "    \n",
    "    Other combinations of the HLA genes can not be used as input.    \n",
    "    \n",
    "    This function returns the estimated allele frequencies/counts and alleles of the sample from Hapl-o-Mat.\n",
    "    '''\n",
    "    \n",
    "    if hla_gene not in ('A', 'B', 'C', 'DRB1', 'DQB1', 'DPB1'):\n",
    "        raise Exception(\"The hla_gene can only be 'A', 'B', 'C', 'DRB1', 'DQB1' or 'DPB1'\")\n",
    "        \n",
    "    if counts and alleles:\n",
    "        raise Exception('Only 1 output can come from this function.')\n",
    "    \n",
    "    with open(hfs_dat, 'r') as f:\n",
    "        f = f.readlines()\n",
    "        \n",
    "    f = [x.strip('\\n').split('\\t') for x in f]\n",
    "    allele_freq = [(x[0].split('~'), float(x[1])) for x in f]\n",
    "    allele_freq = [([y.strip('g') for y in x[0]],x[1]) for x in allele_freq]\n",
    "    \n",
    "    if hla_gene == 'A':\n",
    "        n = 0        \n",
    "    elif hla_gene == 'B':\n",
    "        n = 1        \n",
    "    elif hla_gene == 'C':\n",
    "        n = 2        \n",
    "    \n",
    "    elif hla_gene == 'DRB1': # Hapl-o-Mat orders the genes in the haplotypes alphabetically.\n",
    "        if len(allele_freq[0][0]) == 4:     \n",
    "            n = 3\n",
    "        elif len(allele_freq[0][0]) == 5:     \n",
    "            n = 4    \n",
    "        elif len(allele_freq[0][0]) == 6:\n",
    "            n = 5\n",
    "        else:\n",
    "            raise Exception('The haplotypes should have 4, 5 or 6 genes in a specific order. Read the documentation.')\n",
    "    \n",
    "    elif hla_gene == 'DQB1':\n",
    "        if len(allele_freq[0][0]) == 4:     \n",
    "            raise Exception('DQB1 should not be in a haplotype with only 4 genes. Read the documentation.')\n",
    "        elif len(allele_freq[0][0]) == 5:     \n",
    "            n = 3    \n",
    "        elif len(allele_freq[0][0]) == 6:\n",
    "            n = 4\n",
    "        else:\n",
    "            raise Exception('The haplotypes should have 4, 5 or 6 genes in a specific order. Read the documentation.')                 \n",
    "    \n",
    "    elif hla_gene == 'DPB1':\n",
    "        if len(allele_freq[0][0]) == 4:     \n",
    "            raise Exception('DPB1 should not be in a haplotype with only 4 genes. Read the documentation.')\n",
    "        elif len(allele_freq[0][0]) == 5:     \n",
    "            raise Exception('DPB1 should not be in a haplotype with only 5 genes. Read the documentation.')    \n",
    "        elif len(allele_freq[0][0]) == 6:\n",
    "            n = 3\n",
    "        else:\n",
    "            raise Exception('The haplotypes should have 4, 5 or 6 genes in a specific order. Read the documentation.')\n",
    "    \n",
    "    HLA_gene = {el:0 for el in [x[0][n] for x in allele_freq]}\n",
    "    \n",
    "    for x in allele_freq:\n",
    "        HLA_gene[x[0][n]] += x[1]\n",
    "    \n",
    "    HLA_gene = {k:v for k, v in sorted(HLA_gene.items(), key=lambda item: item[1], reverse=True)}\n",
    "    \n",
    "    if alleles:\n",
    "        return list(HLA_gene.keys())\n",
    "    \n",
    "    elif counts:        \n",
    "        return {k:round(v*num_of_people) for k,v in HLA_gene.items()} \n",
    "        # No need to multiply by 2, because of the 2 genotypes on MAC input\n",
    "    \n",
    "    else:\n",
    "        return HLA_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "aed3b6e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "g4_path = '/Users/vasou/Downloads/test.dat'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "ba6cd0c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_results = af_estimation_from_HaploMat_g_group(g4_path, 'A', 1835) #g4_path = dat file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "d78879a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A*24:02': 0.08969907407407002,\n",
       " 'A*29:25': 0.06770833333333,\n",
       " 'A*11:01': 0.06770833333333,\n",
       " 'A*02:77': 0.06250000000001,\n",
       " 'A*03:217': 0.05729166666666,\n",
       " 'A*23:06': 0.04687500000003,\n",
       " 'A*68:70': 0.046875,\n",
       " 'A*23:01': 0.04600694444447,\n",
       " 'A*02:454': 0.04166666666667,\n",
       " 'A*29:67': 0.036458333333340004,\n",
       " 'A*30:13': 0.03645833333333,\n",
       " 'A*03:02': 0.03125,\n",
       " 'A*66:10': 0.03125,\n",
       " 'A*30:73N': 0.031249999999999993,\n",
       " 'A*33:27': 0.026041666666670002,\n",
       " 'A*32:33': 0.02604166666667,\n",
       " 'A*26:04': 0.026041666666660003,\n",
       " 'A*03:93': 0.020833333333330005,\n",
       " 'A*02:570': 0.02083333333333,\n",
       " 'A*32:73': 0.02083333333333,\n",
       " 'A*02:163': 0.020833333333329998,\n",
       " 'A*68:94N': 0.02083333333332,\n",
       " 'A*02:258': 0.015625,\n",
       " 'A*03:239': 0.01562499999992,\n",
       " 'A*11:109N': 0.01041666666667,\n",
       " 'A*02:259': 0.01041666666667,\n",
       " 'A*02:158': 0.01041666666667,\n",
       " 'A*29:36': 0.01041666666666,\n",
       " 'A*29:01': 0.00520833333333,\n",
       " 'A*23:03': 0.00520833333333,\n",
       " 'A*68:01': 0.00520833333333,\n",
       " 'A*24:215': 0.00520833333333,\n",
       " 'A*24:05': 0.00520833333333,\n",
       " 'A*29:02': 0.00520833333333,\n",
       " 'A*11:151': 0.00520833333333,\n",
       " 'A*02:289': 0.00520833333333,\n",
       " 'A*32:75': 0.00520638700941,\n",
       " 'A*24:50': 0.00405092592592,\n",
       " 'A*23:04': 0.00086805555556}"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
