{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ebf01093",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import  math\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "import matplotlib.pyplot as plt\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b13c4543",
   "metadata": {},
   "source": [
    "## Pipeline for Matching Coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "34e5e14c",
   "metadata": {},
   "outputs": [],
   "source": [
    "###############################################################################################################\n",
    "##                                               PIPELINE                                                    ##\n",
    "###############################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "21360042",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Clean the estimated haplotypes (haplotypes from haplomat=>drop ~ and g/G)\n",
    "'''\n",
    "def clean_data(\n",
    "                filename, #dataframe with haplotypes and frequency,\n",
    "                n, #number of samples that the haplotypes derived\n",
    "              ):\n",
    "    '''\n",
    "    Read Estimated Haplotypes from Haplomat\n",
    "    '''\n",
    "    if 'DKMS' in filename and '3loci' in filename:\n",
    "        hap_estimated = pd.read_csv(f'/Users/vasou/Documents/HLA/Matching Coverage/Estimated_Haplotypes/{filename}.dat',\n",
    "                                    sep = '\\t',\n",
    "                                    decimal=',',\n",
    "                                    header = None)\n",
    "    else:\n",
    "        hap_estimated = pd.read_csv(f'/Users/vasou/Documents/HLA/Matching Coverage/Estimated_Haplotypes/{filename}.dat',\n",
    "                                    sep = '\\t',\n",
    "                                    header = None)\n",
    "        \n",
    "    #hap_estimated = pd.read_csv(f'/Users/vasou/Documents/GitHub/HLA.github.io/Analyses/Haplotype_Analyses/Haplotype_data/{filename}.dat', sep=\"\\t\",header=None)\n",
    "    hap_estimated.columns = ['Haplotypes', 'Frequency']\n",
    "    #display(hap_estimated)\n",
    "    print(f'previous len: {hap_estimated.shape[0]}')\n",
    "    correction = 1/(2*n)\n",
    "    hap_estimated = hap_estimated[hap_estimated['Frequency']>=correction]\n",
    "    \n",
    "    #We want the estimated haplotpyes have >=2 counts (second correction)\n",
    "    counts = [round(freq*n) for freq in hap_estimated['Frequency'].values]\n",
    "   \n",
    "    hap_estimated['Counts'] = counts\n",
    "    counts_correction = 2\n",
    "    #assert sum(counts) == n\n",
    "    hap_estimated = hap_estimated[hap_estimated['Counts']>=counts_correction]\n",
    "    print(f'samples: {sum(hap_estimated[\"Counts\"].values)}')\n",
    "    haplotypes = []\n",
    "    for haplotype in hap_estimated['Haplotypes'].values:\n",
    "        splited_haplotype = haplotype.split('~')\n",
    "        splited_haplotype = [splited_haplotype[i][:-1] if splited_haplotype[i][-1].isalpha() else splited_haplotype[i] for i in range(len(splited_haplotype)) ]\n",
    "        haplotype = ' '.join(splited_haplotype)\n",
    "        haplotypes.append(haplotype)\n",
    "    #display(hap_estimated)\n",
    "    print(f'After len: {hap_estimated.shape[0]}')\n",
    "    return haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "204235b4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "90e04c1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Calculate matching coverage\n",
    "'''\n",
    "\n",
    "def matching_coverage(\n",
    "                        dat_data,\n",
    "                        haplotypes, #estimated haplotypes from haplomat or haplotypes from homozygous\n",
    "                        genes,\n",
    "                        not_found = False\n",
    "                    ):\n",
    "    not_found_ids = []\n",
    "    \n",
    "    if not_found:\n",
    "        if genes == 3:\n",
    "            case = ['ID','A1', 'A2','B1','B2','DRB1_1', 'DRB1_2']\n",
    "            for index, row in dat_data[case].iterrows():\n",
    "                if row['A1'] == 'not found' and row['A2'] == 'not found':\n",
    "                    not_found_ids.append(index)\n",
    "                elif row['B1'] == 'not found' and row['B2'] == 'not found':\n",
    "                    not_found_ids.append(index)\n",
    "                elif row['DRB1_1'] == 'not found' and row['DRB1_2'] == 'not found':\n",
    "                    not_found_ids.append(index)\n",
    "        else:\n",
    "            case = ['ID','A1', 'A2','B1','B2','C1','C2','DRB1_1', 'DRB1_2','DQB1_1', 'DQB1_2']\n",
    "            for index, row in dat_data[case].iterrows():\n",
    "                if row['A1'] == 'not found' and row['A2'] == 'not found':\n",
    "                    not_found_ids.append(index)\n",
    "                elif row['B1'] == 'not found' and row['B2'] == 'not found':\n",
    "                    not_found_ids.append(index)\n",
    "                elif row['C1'] == 'not found' and row['C2'] == 'not found':\n",
    "                    not_found_ids.append(index)\n",
    "                elif row['DRB1_1'] == 'not found' and row['DRB1_2'] == 'not found':\n",
    "                    not_found_ids.append(index)\n",
    "                elif row['DQB1_1'] == 'not found' and row['DQB1_2'] == 'not found':\n",
    "                    not_found_ids.append(index)\n",
    "    else:\n",
    "        if genes == 3:\n",
    "            case = ['ID','A1', 'A2','B1','B2','DRB1_1', 'DRB1_2']\n",
    "        else:\n",
    "            case = ['ID','A1', 'A2','B1','B2','C1','C2','DRB1_1', 'DRB1_2','DQB1_1', 'DQB1_2']\n",
    "            \n",
    "        for index, row in dat_data[case].iterrows():\n",
    "                sum_row = sum([1 for allele in case[1:] if row[allele]=='not found'])\n",
    "                if sum_row > 0:\n",
    "                    not_found_ids.append(index)\n",
    "        \n",
    "        \n",
    "    #data_nf = dat_data[case][~dat_data[case].apply(lambda row: row.astype(str).str.contains('not found').any(), axis=1)]\n",
    "    dat_data.drop(labels = not_found_ids, axis = 0, inplace = True)\n",
    "            \n",
    "    #Add the loci before allele: A*01:01\n",
    "    data = dat_data[case].copy()\n",
    "    for column in data.columns[1:]: #the first column is the ID of each sample\n",
    "        l = []\n",
    "        for value in data[column].values:\n",
    "            if column in ['DRB1_1', 'DRB1_2','DQB1_1','DQB1_2','DPB1_1','DPB1_2']:\n",
    "                l.append(column.split('_')[0] + '*' + ':'.join(value.split(':')[:2]))\n",
    "            else: \n",
    "                l.append(column[0] + '*' + ':'.join(value.split(':')[:2]))\n",
    "\n",
    "        data[column] = l\n",
    "    #data.to_excel('test.xlsx')  \n",
    "    display(data)\n",
    "    count_hap = defaultdict(lambda:0)\n",
    "    ids = defaultdict(lambda:[])\n",
    "    remaining_data = data.copy()\n",
    "    for top in haplotypes:\n",
    "        unique_hap = top.split(' ')\n",
    "        \n",
    "        for index, row in remaining_data.iterrows():\n",
    "            haplotype = []\n",
    "            if unique_hap[0] in row['A1'] or unique_hap[0] in row['A2']:\n",
    "                haplotype.append(unique_hap[0])\n",
    "\n",
    "            if unique_hap[1] in row['B1'] or unique_hap[1] in row['B2']:\n",
    "                haplotype.append(unique_hap[1])\n",
    "                \n",
    "            if genes ==3:\n",
    "                if unique_hap[2] in row['DRB1_1'] or unique_hap[2] in row['DRB1_2']:\n",
    "                    haplotype.append(unique_hap[2])\n",
    "                \n",
    "            if genes == 5:\n",
    "                if unique_hap[2] in row['C1'] or unique_hap[2] in row['C2']:\n",
    "                    haplotype.append(unique_hap[2])\n",
    "                if unique_hap[3] in row['DQB1_1'] or unique_hap[3] in row['DQB1_2']:\n",
    "                    haplotype.append(unique_hap[3])\n",
    "                if unique_hap[4] in row['DRB1_1'] or unique_hap[4] in row['DRB1_2']:\n",
    "                    haplotype.append(unique_hap[4])  \n",
    "\n",
    "            haplotype = ' '.join(haplotype)\n",
    "            #print(haplotype)\n",
    "            #print('-------------------------------------------')\n",
    "            #print(top)\n",
    "            #print('=======================================')\n",
    "            if haplotype == top:\n",
    "                count_hap[haplotype]+=1\n",
    "                ids[haplotype].append((remaining_data['ID'][index],index))\n",
    "                #print(haplotype)\n",
    "        print(remaining_data.shape)\n",
    "        remaining_data.drop(labels = [i[1] for i in ids[top]], axis = 0, inplace = True)\n",
    "        print(remaining_data.shape)\n",
    "        \n",
    "    coverage = defaultdict()\n",
    "    n = data.shape[0]#dat_data.shape[0]\n",
    "    print(n)\n",
    "    if n==0:\n",
    "        n=0.0001\n",
    "    for key in count_hap:\n",
    "        coverage[key] = count_hap[key]/n\n",
    "            \n",
    "    return coverage, ids, count_hap\n",
    "\n",
    "#matching_coverage,ids, count_hap = matching_coverage(data_4451, est_hap, genes = 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "cb520004",
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################\n",
    "##          DATA           ##\n",
    "#############################\n",
    "greece_CBUs_BMDs_2fields_3loci = pd.read_excel(f'/Users/vasou/Documents/HLA/Matching Coverage/MC_for_paper/Data/CBUs_BMDs_for_Hap_from_80804.xlsx', \n",
    "                                               sheet_name = 'CBUs&BMDs_2fields_3loci_78611',\n",
    "                                               header = [0])\n",
    "greece_CBUs_BMDs_2fields_5loci = pd.read_excel(f'/Users/vasou/Documents/HLA/Matching Coverage/MC_for_paper/Data/CBUs_BMDs_for_Hap_from_80804.xlsx',\n",
    "                                               sheet_name = 'CBUs&BMDs_2fields_5loci_70222',\n",
    "                                               header = [0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59ff2ccd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#########################\n",
    "##         RUN         ##\n",
    "#########################\n",
    "with pd.ExcelWriter('MC_2_2fields_Greece_12Countries_BMDs_to_Greece_all.xlsx', engine = \"openpyxl\", mode = \"w\") as writer:\n",
    "    for n_gene in [3,5]:\n",
    "        if n_gene == 5:\n",
    "            est_hap = clean_data('Cyprus/RunX_htf_cyprus_cbus_2fields_2702_5loci', 2702)\n",
    "            data = greece_CBUs_BMDs_2fields_5loci\n",
    "            if 'BIRTH' in data.columns:\n",
    "                data.columns = ['ID','BIRTH', 'A1', 'A2','B1','B2','C1','C2','DRB1_1','DRB1_2','DQB1_1','DQB1_2']\n",
    "            else:\n",
    "                data.columns = ['ID', 'A1', 'A2','B1','B2','C1','C2','DRB1_1','DRB1_2','DQB1_1','DQB1_2']\n",
    "        else:\n",
    "            est_hap = clean_data('Cyprus/RunX_htf_cyprus_cbus_2fields_2843_3loci', 2843)\n",
    "            data = greece_CBUs_BMDs_2fields_3loci\n",
    "            if 'BIRTH' in data.columns:\n",
    "                data.columns = ['ID','BIRTH', 'A1', 'A2','B1','B2','DRB1_1', 'DRB1_2']\n",
    "            else:\n",
    "                data.columns = ['ID','A1', 'A2','B1','B2','DRB1_1', 'DRB1_2']\n",
    "                \n",
    "            \n",
    "        matching_cov, ids, count_hap = matching_coverage(data, est_hap, genes = n_gene, not_found = False)\n",
    "        #not_found = True if allow not found\n",
    "            \n",
    "        a = pd.DataFrame(\n",
    "            [list(count_hap.keys()), list(count_hap.values()), list(matching_cov.values())],\n",
    "            ).T\n",
    "        a.columns = ['Haplotype','Counts', 'Coverage']\n",
    "        a.to_excel(writer, sheet_name = f'{n_gene}_genes', engine = \"openpyxl\", index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
