{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "163bad19",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################################\n",
    "##              Import libraries                 ##\n",
    "###################################################\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "from collections import defaultdict\n",
    "import  math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b236acf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################################\n",
    "##                        Read the files                            ##\n",
    "######################################################################\n",
    "data = {}\n",
    "for file in ['chania','rethymno','heraklion','lasithi']:\n",
    "    data[file] = pd.read_excel(f'/Users/vasou/Documents/HLA/DHTOB/{file}.xlsx')\n",
    "    \n",
    "#all_cretans = pd.read_excel(f'/Users/vasou/Documents/HLA/1744_COMPLETE_CRETANS.xlsx')\n",
    "#all_cretans.rename(columns = {'#':'id'}, inplace = True)\n",
    "#all_cretans.replace('?',np.NaN, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e847daf6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d10f1cf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "##############################################################################  \n",
    "##                 Found the alleles for all regions                        ##\n",
    "##############################################################################\n",
    "\n",
    "def unique_and_sorted_allele(data, #Dictionary: key (string) = region (chania,rethymno etc), value(dataframe) = allleles\n",
    "                            ):\n",
    "    \n",
    "    alleles_all_regions = defaultdict(list)\n",
    "    \n",
    "    for region in data:\n",
    "        for allele in ['A', 'B', 'C', 'DRB1','DQB1','DPB1']:#set([i.split('_')[0] if '_' in i else i[0] for i in data[key].columns[1:]]):\n",
    "            alleles_all_regions[allele].append(list(pd.unique(\n",
    "                                                data[region][[allele+'_1' if 'D' in allele else allele + '1',\n",
    "                                                allele+'_2' if 'D' in allele else allele + '2']].values.ravel())))\n",
    "\n",
    "    alleles_all_regions = {allele:sorted(set(list(itertools.chain(*alleles_all_regions[allele])))) for allele in alleles_all_regions}\n",
    "    \n",
    "    return alleles_all_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0ee2460",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5958a2db",
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################################################################################  \n",
    "##                            Allele and Genotype counts                                   ##\n",
    "#############################################################################################\n",
    "\n",
    "\n",
    "def allele_genotype_counts(data, #Dictionary: key = region(chania,rethymno etc), value(dataframe) = allleles\n",
    "                           type_count = 'Allele' #Allele: returns allele counts, Genotype:returns observed genotype counts\n",
    "                          ):\n",
    "    \n",
    "    count_dict = defaultdict(lambda:defaultdict())\n",
    "    \n",
    "    ########################################################\n",
    "    ##   First found the unique alleles from all regions  ##\n",
    "    ########################################################\n",
    "    alleles_all_regions = unique_and_sorted_allele(data)\n",
    "\n",
    "    for region in data:\n",
    "        for allele in alleles_all_regions:\n",
    "            column_1 = allele + '_1' if 'D' in allele else allele + '1'\n",
    "            column_2 = allele + '_2' if 'D' in allele else allele + '2'\n",
    "            total_het = 0\n",
    "            total_hom = 0\n",
    "            for i in range(len(alleles_all_regions[allele])):\n",
    "\n",
    "                #Check Hom\n",
    "                count_hom = sum((data[region][column_1] == alleles_all_regions[allele][i]) & (\n",
    "                            data[region][column_2] == alleles_all_regions[allele][i]))\n",
    "                \n",
    "                if type_count == 'Allele': \n",
    "                    count_hom = 2*count_hom\n",
    "                    \n",
    "                    #Het\n",
    "                    count_het_1 = sum((data[region][column_1] == alleles_all_regions[allele][i]))\n",
    "                    count_het_2 = sum((data[region][column_2] == alleles_all_regions[allele][i]))\n",
    "                    count_het = (count_het_1 + count_het_2) - count_hom\n",
    "                else:\n",
    "                    #Het\n",
    "                    count_het = sum((data[region][column_1] == alleles_all_regions[allele][i])) - count_hom\n",
    "                    \n",
    "                total_het += count_het\n",
    "                total_hom += count_hom\n",
    "                \n",
    "                count_dict[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]] = count_hom\n",
    "                \n",
    "                #Store Het\n",
    "                count_dict[f'het_{region}_{allele}'][alleles_all_regions[allele][i]] = count_het\n",
    "                \n",
    "            count_dict[f'n_hom_{allele}_{region}'] = total_hom\n",
    "            count_dict[f'n_het_{allele}_{region}'] = total_het\n",
    "                        \n",
    "\n",
    "            #for i in range(len(alleles_all_regions[allele])):\n",
    "            #    \n",
    "            #    count_dict[f'het_rest_{region}_{allele}'][alleles_all_regions[allele][i]] = total_het - count_dict[f'het_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "            #    count_dict[f'hom_rest_{region}_{allele}'][alleles_all_regions[allele][i]] = total_hom - count_dict[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "    \n",
    "    \n",
    "    for allele in alleles_all_regions:\n",
    "        hom_total = 0\n",
    "        het_total = 0\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "            hom = 0\n",
    "            het = 0\n",
    "            for region in data:\n",
    "                hom += count_dict[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "                het += count_dict[f'het_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "                \n",
    "                 \n",
    "                count_dict[f'het_rest_{region}_{allele}'][alleles_all_regions[allele][i]] = count_dict[f'n_het_{allele}_{region}'] - count_dict[f'het_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "                count_dict[f'hom_rest_{region}_{allele}'][alleles_all_regions[allele][i]] = count_dict[f'n_hom_{allele}_{region}'] - count_dict[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "\n",
    "                    \n",
    "            count_dict[f'hom_total_{allele}'][alleles_all_regions[allele][i]] = hom \n",
    "            count_dict[f'het_total_{allele}'][alleles_all_regions[allele][i]] = het\n",
    "            hom_total += hom\n",
    "            het_total += het\n",
    "            \n",
    "        count_dict[f'n_hom_total_{allele}'] = hom_total\n",
    "        count_dict[f'n_het_total_{allele}'] = het_total\n",
    "            \n",
    "\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "\n",
    "            count_dict[f'hom_rest_total_{allele}'][alleles_all_regions[allele][i]] = hom_total - count_dict[f'hom_total_{allele}'][alleles_all_regions[allele][i]]\n",
    "            count_dict[f'het_rest_total_{allele}'][alleles_all_regions[allele][i]] = het_total - count_dict[f'het_total_{allele}'][alleles_all_regions[allele][i]]\n",
    "           \n",
    "    return count_dict\n",
    "\n",
    "\n",
    "#allele_count = allele_genotype_counts(data, type_count = 'Allele')\n",
    "#obs_gen_count = allele_genotype_counts(data, type_count = 'Genotype')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d8fed4ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##                         Step 1: Allele Frequencies                                ##\n",
    "#######################################################################################\n",
    "\n",
    "\n",
    "def help_func_freq(a,b,c,d, allele_freq, key_part1, key_part2):\n",
    "    nominator = a +b\n",
    "    denominator = c + d\n",
    "    calc_p = nominator/denominator\n",
    "    \n",
    "    allele_freq[f'p_{key_part1}'][key_part2] = calc_p\n",
    "    allele_freq[f'p^2_{key_part1}'][key_part2] = calc_p * calc_p\n",
    "    allele_freq[f'q_{key_part1}'][key_part2] = 1 - calc_p\n",
    "    allele_freq[f'q^2_{key_part1}'][key_part2] = (1 - calc_p)**2\n",
    "    \n",
    "    \n",
    "def allele_frequencies(data #Dictionary: key (string) = region (chania,rethymno etc), value(dataframe) = allleles\n",
    "                      ):\n",
    "    \n",
    "    n_allele = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    allele_freq = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    \n",
    "    \n",
    "    ########################################################\n",
    "    ##   First found the unique alleles from all regions  ##\n",
    "    ########################################################\n",
    "    alleles_all_regions = unique_and_sorted_allele(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##                   Allele counts                    ##\n",
    "    ########################################################\n",
    "    allele_count = allele_genotype_counts(data, type_count = 'Allele')\n",
    "\n",
    "    ########################################################\n",
    "    ##              Calculate frequencies                 ##\n",
    "    ########################################################\n",
    "\n",
    "    for allele in alleles_all_regions:\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "            total_samples = 0\n",
    "            for region in data:\n",
    "                total_samples += data[region].shape[0]\n",
    "                help_func_freq(allele_count[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                          allele_count[f'het_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                          allele_count[f'n_hom_{allele}_{region}'],\n",
    "                          allele_count[f'n_het_{allele}_{region}'], \n",
    "                          allele_freq, \n",
    "                          f\"{region}_{allele}\", \n",
    "                          alleles_all_regions[allele][i])\n",
    "                \n",
    "               \n",
    "            help_func_freq(allele_count[f'hom_total_{allele}'][alleles_all_regions[allele][i]],\n",
    "                      allele_count[f'het_total_{allele}'][alleles_all_regions[allele][i]],\n",
    "                      allele_count[f'n_hom_total_{allele}'], \n",
    "                      allele_count[f'n_het_total_{allele}'],\n",
    "                      allele_freq,\n",
    "                      f\"total_{allele}\", \n",
    "                      alleles_all_regions[allele][i])\n",
    "\n",
    "\n",
    "    return allele_freq\n",
    "\n",
    "\n",
    "#allele_freq = allele_frequencies(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "54830c4c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##             Step 2: expected genotypic counts (under HWE)                         ##\n",
    "#######################################################################################\n",
    "\n",
    "\n",
    "def help_func_hwe(a,b,c,d,counts_HWE, key_part1, key_part2):\n",
    "    counts_HWE[f'hom_p^2_{key_part1}'][key_part2] = a*b\n",
    "    counts_HWE[f'het_2pq_{key_part1}'][key_part2] = 2*c*d*b\n",
    "\n",
    "def HWE(data):\n",
    "    counts_HWE = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    \n",
    "    ########################################################\n",
    "    ##   First found the unique alleles from all regions  ##\n",
    "    ########################################################\n",
    "    alleles_all_regions = unique_and_sorted_allele(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##            Calculate allele frequencies            ##\n",
    "    ########################################################\n",
    "    allele_freq = allele_frequencies(data)\n",
    "    \n",
    "    for allele in alleles_all_regions:\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "            total_samples = 0\n",
    "            for region in data:\n",
    "                total_samples += data[region].shape[0]\n",
    "                \n",
    "                help_func_hwe(allele_freq[f'p^2_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              data[region].shape[0],\n",
    "                              allele_freq[f'p_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              allele_freq[f'q_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              counts_HWE,\n",
    "                              f'{region}_{allele}',\n",
    "                              alleles_all_regions[allele][i] \n",
    "                             )\n",
    "                \n",
    "            help_func_hwe(allele_freq[f'p^2_total_{allele}'][alleles_all_regions[allele][i]],\n",
    "                          total_samples,\n",
    "                          allele_freq[f'p_total_{allele}'][alleles_all_regions[allele][i]],\n",
    "                          allele_freq[f'q_total_{allele}'][alleles_all_regions[allele][i]],\n",
    "                          counts_HWE,\n",
    "                          f'total_{allele}',\n",
    "                          alleles_all_regions[allele][i] \n",
    "                         )\n",
    "    return counts_HWE\n",
    "              \n",
    "#counts_HWE = HWE(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85db5f75",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "86a8833f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##                 Step 3: excess or deficiencies of observed                        ##\n",
    "##      homozygotes in each subpopulation relative to HWE genotypic counts           ##\n",
    "#######################################################################################\n",
    "\n",
    "def deficiencies_sub(data):\n",
    "    obs_exp = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    \n",
    "    ########################################################\n",
    "    ##   First found the unique alleles from all regions  ##\n",
    "    ########################################################\n",
    "    alleles_all_regions = unique_and_sorted_allele(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##                   Allele counts                    ##\n",
    "    ########################################################\n",
    "    obs_gen_count = allele_genotype_counts(data, type_count = 'Genotype')\n",
    "    \n",
    "    ########################################################\n",
    "    ##                   Calculate HWE                    ##\n",
    "    ########################################################\n",
    "    counts_HWE = HWE(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##             Calculate deficiencies                 ##\n",
    "    ########################################################\n",
    "    for allele in alleles_all_regions:\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "            total_samples = 0\n",
    "            for region in data:\n",
    "    \n",
    "                obs_exp[f'obs-exp_p^2_{region}_{allele}'][alleles_all_regions[allele][i]] = (\n",
    "                                obs_gen_count[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]] -\n",
    "                                counts_HWE[f'hom_p^2_{region}_{allele}'][alleles_all_regions[allele][i]])\n",
    "            obs_exp[f'obs-exp_p^2_total_{allele}'][alleles_all_regions[allele][i]] = (\n",
    "                            obs_gen_count[f'hom_total_{allele}'][alleles_all_regions[allele][i]]-\n",
    "                            counts_HWE[f'hom_p^2_total_{allele}'][alleles_all_regions[allele][i]])\n",
    "\n",
    "    return obs_exp\n",
    "\n",
    "\n",
    "#excess = deficiencies_sub(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0e8c84af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##   Step4: excess or deficiencies of observed homozygotes in the total population   ##\n",
    "#######################################################################################\n",
    "\n",
    "def deficiencies_total(data):\n",
    "    sum_obs_exp = defaultdict(lambda: 0)\n",
    "\n",
    "    ########################################################\n",
    "    ##          Calculate deficiencies_sub                ##\n",
    "    ########################################################\n",
    "    excess = deficiencies_sub(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##             Calculate deficiencies                 ##\n",
    "    ########################################################\n",
    "    for key in excess:\n",
    "\n",
    "        sum_obs_exp[f'sum_{key}'] = round(sum(excess[key].values()))\n",
    "\n",
    "    return sum_obs_exp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "828fa900",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##             Step 5: local observed heterozygosities of each subpopulation         ##\n",
    "#######################################################################################\n",
    "\n",
    "\n",
    "def obs_heterozygosities(data):\n",
    "    obs_het = defaultdict(lambda:0)\n",
    "    \n",
    "    ########################################################\n",
    "    ##                   Allele counts                    ##\n",
    "    ########################################################\n",
    "    obs_gen_count = allele_genotype_counts(data, type_count = 'Genotype')\n",
    "    \n",
    "    ########################################################\n",
    "    ##           local observed heterozygosities          ##\n",
    "    ########################################################\n",
    "    \n",
    "    for allele in data['chania'].columns[1:]:\n",
    "        total_samples = 0\n",
    "        for region in data:\n",
    "            total_samples+=data[region].shape[0]\n",
    "            allele = allele.split('_')[0] if 'D' in allele else allele[0]\n",
    "            obs_het[f'Het_{region}_{allele}'] = obs_gen_count[f'n_het_{allele}_{region}']/data[region].shape[0]\n",
    "        obs_het[f'Het_total_{allele}'] = obs_gen_count[f'n_het_total_{allele}']/total_samples\n",
    "            \n",
    "    return obs_het\n",
    "\n",
    "\n",
    "#obs_het = obs_heterozygosities(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d9c4451a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##               Step 6: expected heterozygosities of each subpopulation             ##\n",
    "#######################################################################################\n",
    "\n",
    "def exp_heterozygosities(data):\n",
    "    exp_het = defaultdict(lambda:0)\n",
    "    ########################################################\n",
    "    ##            Calculate allele frequencies            ##\n",
    "    ########################################################\n",
    "    allele_freq = allele_frequencies(data)\n",
    "    \n",
    "    for allele in data['chania'].columns[1:]:\n",
    "        allele = allele.split('_')[0] if 'D' in allele else allele[0]\n",
    "        for region in data:\n",
    "            exp_het[f'Het_{region}_{allele}'] = 1 - sum(allele_freq[f'p^2_{region}_{allele}'].values())\n",
    "        exp_het[f'Het_total_{allele}'] = 1 - sum(allele_freq[f'p^2_total_{allele}'].values())\n",
    "    return exp_het\n",
    "\n",
    "\n",
    "#expected_het = exp_heterozygosities(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e469b490",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##  Step 7: local inbreeding coefficient of each subpopulation (Fs=Hexp-Hobs/Hexp)   ##\n",
    "#######################################################################################\n",
    "\n",
    "def Fs_coefficient(data):\n",
    "    \n",
    "    Fs = defaultdict(lambda: 0)\n",
    "    \n",
    "    ########################################################\n",
    "    ##         Calculate expected heterozygosities        ##\n",
    "    ########################################################\n",
    "    expected_het = exp_heterozygosities(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##         Calculate observed heterozygosities        ##\n",
    "    ########################################################\n",
    "    obs_het = obs_heterozygosities(data)\n",
    "    \n",
    "    for allele in data['chania'].columns[1:]:\n",
    "        allele = allele.split('_')[0] if 'D' in allele else allele[0]\n",
    "        for region in data:\n",
    "            Fs[f'Fs_{region}_{allele}'] = (expected_het[f'Het_{region}_{allele}'] - \n",
    "                                           obs_het[f'Het_{region}_{allele}'])/(\n",
    "                                           expected_het[f'Het_{region}_{allele}'])\n",
    "            \n",
    "        Fs[f'Fs_total_{allele}'] = (expected_het[f'Het_total_{allele}'] - \n",
    "                                           obs_het[f'Het_total_{allele}'])/(\n",
    "                                            expected_het[f'Het_total_{allele}'])\n",
    "            \n",
    "    return Fs\n",
    "            \n",
    "\n",
    "#fs_coef = Fs_coefficient(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b457008f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##            Step8: p_allele bar (global frequency of each allele                   ##\n",
    "##       across all subpopulations and weighted by each subpopulation size)          ##\n",
    "#######################################################################################\n",
    "\n",
    "\n",
    "def help_func_bar(a,b,c,d,p_allele_bar, key_part1, key_part2):\n",
    "    \n",
    "    if d == 0:\n",
    "        denominator = b\n",
    "        nominator = c*a\n",
    "        p_allele_bar[f'{key_part1}'][key_part2] += nominator/denominator\n",
    "        p_allele_bar[f'{key_part1}_square'][key_part2] = p_allele_bar[f'{key_part1}'][key_part2]**2\n",
    "    else: \n",
    "        denominator = a+b\n",
    "        nominator = c*a + d*b\n",
    "        p_allele_bar[f'{key_part1}'][key_part2] = nominator/denominator\n",
    "        p_allele_bar[f'{key_part1}_square'][key_part2] = (nominator/denominator)**2\n",
    "\n",
    "def p_allele_bar(data):\n",
    "    n_allele = defaultdict(lambda: 0)\n",
    "    p_allele_bar = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    \n",
    "    ########################################################\n",
    "    ##   First found the unique alleles from all regions  ##\n",
    "    ########################################################\n",
    "    alleles_all_regions = unique_and_sorted_allele(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##            Calculate allele frequencies            ##\n",
    "    ########################################################\n",
    "    allele_freq = allele_frequencies(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##                   Allele counts                    ##\n",
    "    ########################################################\n",
    "    allele_count = allele_genotype_counts(data, type_count = 'Allele')\n",
    "    \n",
    "    total_samples = sum([data[region].shape[0] for region in data])*2\n",
    "    for allele in alleles_all_regions:\n",
    "        for region in data:\n",
    "            for i in range(len(alleles_all_regions[allele])):\n",
    "                \n",
    "                help_func_bar(data[region].shape[0]*2,\n",
    "                              total_samples,\n",
    "                              allele_freq[f'p_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              0,\n",
    "                              p_allele_bar,\n",
    "                              f'all_{allele}',\n",
    "                              alleles_all_regions[allele][i]\n",
    "                             )\n",
    "\n",
    "\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "            for region_1, region_2 in list(itertools.combinations(data.keys(), 2)):\n",
    "\n",
    "                help_func_bar(data[region_1].shape[0]*2,\n",
    "                              data[region_2].shape[0]*2,\n",
    "                              allele_freq[f'p_{region_1}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              allele_freq[f'p_{region_2}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              p_allele_bar,\n",
    "                              f'{region_1}_{region_2}_{allele}',\n",
    "                              alleles_all_regions[allele][i]\n",
    "                             )\n",
    "\n",
    "    return p_allele_bar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "36ff19ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##                   Step9&10&11: heterozygosity indices                             ##\n",
    "##              (Hi=weighted average of observed heterozygosities),                  ##\n",
    "##              (Hs =weighted average of expected heterozygosities),                 ##\n",
    "## (Ht=expected hetrozygosty based on the global gene frequencies i.e. p_allele-bar) ##\n",
    "#######################################################################################\n",
    "\n",
    "def heterozygosity_indices(data):\n",
    "    \n",
    "    Het_indices = defaultdict(lambda: 0)\n",
    "    \n",
    "    ########################################################\n",
    "    ##         Calculate expected heterozygosities        ##\n",
    "    ########################################################\n",
    "    expected_het = exp_heterozygosities(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##         Calculate observed heterozygosities        ##\n",
    "    ########################################################\n",
    "    obs_het = obs_heterozygosities(data)\n",
    "\n",
    "    ########################################################\n",
    "    ##         Calculate observed heterozygosities        ##\n",
    "    ########################################################\n",
    "    allele_bar = p_allele_bar(data)\n",
    "    \n",
    "    \n",
    "    for allele in data['chania'].columns[1:]:\n",
    "        total_samples = 0\n",
    "        allele = allele.split('_')[0] if 'D' in allele else allele[0]\n",
    "        for region in data:\n",
    "            total_samples+=data[region].shape[0]\n",
    "            Het_indices[f'Hi_all_{allele}'] += (obs_het[f'Het_{region}_{allele}'] * data[region].shape[0])\n",
    "            Het_indices[f'Hs_all_{allele}'] += (expected_het[f'Het_{region}_{allele}'] * data[region].shape[0])\n",
    "            \n",
    "        Het_indices[f'Hi_all_{allele}'] /= total_samples\n",
    "        Het_indices[f'Hs_all_{allele}'] /= total_samples\n",
    "        Het_indices[f'Ht_all_{allele}'] = 1 - sum(allele_bar[f'all_{allele}_square'].values())\n",
    "        \n",
    "\n",
    "        for region_1, region_2 in list(itertools.combinations(data.keys(), 2)):  \n",
    "            Het_indices[f'Hi_{region_1}_{region_2}_{allele}'] = (\n",
    "                                                    (obs_het[f'Het_{region_1}_{allele}'] * data[region_1].shape[0])\n",
    "                                                     + (obs_het[f'Het_{region_2}_{allele}'] * data[region_2].shape[0]))\\\n",
    "                                                     / (data[region_1].shape[0] + data[region_2].shape[0])\n",
    "\n",
    "            Het_indices[f'Hs_{region_1}_{region_2}_{allele}'] = (\n",
    "                                                    (expected_het[f'Het_{region_1}_{allele}'] * data[region_1].shape[0])\n",
    "                                                 + (expected_het[f'Het_{region_2}_{allele}'] * data[region_2].shape[0]))\\\n",
    "                                                 / (data[region_1].shape[0] + data[region_2].shape[0])\n",
    "            Het_indices[f'Ht_{region_1}_{region_2}_{allele}'] = 1 - \\\n",
    "                                    sum(allele_bar[f'{region_1}_{region_2}_{allele}_square'].values())\n",
    "            \n",
    "    return Het_indices\n",
    " \n",
    "    \n",
    "    \n",
    "Het_indices = heterozygosity_indices(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "3c55700a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##           Step12: global Fstatistics (FIS_observed heterozygosities,              ## \n",
    "##                    FST_expected heterozygosities,                                 ##\n",
    "##                    FIT_expected heterozygosities)                                 ##\n",
    "#######################################################################################\n",
    "\n",
    "def help_func_Fstats(a,b,c, Fstatistics, key_part):\n",
    "    \n",
    "    Fstatistics[f'Fis_{key_part}'] = (a-b)/a\n",
    "    Fstatistics[f'Fst_{key_part}'] = (c-a)/c\n",
    "    Fstatistics[f'Fit_{key_part}'] = (c-b)/c\n",
    "\n",
    "def global_Fstatistics(data):\n",
    "    \n",
    "    Fstatistics = defaultdict(lambda:0)\n",
    "    \n",
    "    ########################################################\n",
    "    ##          Calculate heterozygosity indices          ##\n",
    "    ########################################################\n",
    "    Het_indices = heterozygosity_indices(data)\n",
    "    \n",
    "    for allele in data['chania'].columns[1:]:\n",
    "        allele = allele.split('_')[0] if 'D' in allele else allele[0]\n",
    "        help_func_Fstats(Het_indices[f'Hs_all_{allele}'],\n",
    "                         Het_indices[f'Hi_all_{allele}'],\n",
    "                         Het_indices[f'Ht_all_{allele}'],\n",
    "                         Fstatistics,\n",
    "                         f'all_{allele}',\n",
    "                         )\n",
    "        \n",
    "        for region_1, region_2 in list(itertools.combinations(data.keys(), 2)):\n",
    "            \n",
    "            help_func_Fstats(Het_indices[f'Hs_{region_1}_{region_2}_{allele}'],\n",
    "                             Het_indices[f'Hi_{region_1}_{region_2}_{allele}'],\n",
    "                             Het_indices[f'Ht_{region_1}_{region_2}_{allele}'],\n",
    "                             Fstatistics,\n",
    "                             f'_{region_1}_{region_2}_{allele}',\n",
    "                            )\n",
    "    return Fstatistics\n",
    "    \n",
    "        \n",
    "#Fstatistics = global_Fstatistics(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ace622e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52c9668a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##  Step4: excess or deficiencies of observed homozygotes in the total population\n",
    "#######################################################################################\n",
    "\n",
    "\n",
    "sum_obs-exp_p^2_Ch = 8 ????????\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7b6ea2b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1941590e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba5b495a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2163cdb1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "943046ef",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "817c8d51",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "f1f75f92",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "cd2bb088",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f389e94",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b95ed654",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "50168f0f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "498ddc29",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f64933d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56fce040",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "4ab91e9a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3f46057",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14b8911a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "643853b7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc6262bc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d944ee6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81d3ef48",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcc50685",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01934a14",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
