{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "163bad19",
   "metadata": {},
   "outputs": [],
   "source": [
    "###################################################\n",
    "##              Import libraries                 ##\n",
    "###################################################\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "from collections import defaultdict\n",
    "import  math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b236acf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################################################\n",
    "##                        Read the files                            ##\n",
    "######################################################################\n",
    "data = {}\n",
    "for file in ['chania','rethymno','heraklion','lasithi']:\n",
    "    data[file] = pd.read_excel(f'/Users/vasou/Documents/HLA/DHTOB/{file}.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d10f1cf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "##############################################################################  \n",
    "##                 Found the alleles for all regions                        ##\n",
    "##############################################################################\n",
    "\n",
    "def unique_and_sorted_allele(data, #Dictionary: key (string) = region (chania,rethymno etc), value(dataframe) = allleles\n",
    "                            ):\n",
    "    \n",
    "    alleles_all_regions = defaultdict(list)\n",
    "    \n",
    "    for region in data:\n",
    "        for allele in ['A', 'B', 'C', 'DRB1','DQB1','DPB1']:#set([i.split('_')[0] if '_' in i else i[0] for i in data[key].columns[1:]]):\n",
    "            alleles_all_regions[allele].append(list(pd.unique(\n",
    "                                                data[region][[allele+'_1' if 'D' in allele else allele + '1',\n",
    "                                                allele+'_2' if 'D' in allele else allele + '2']].values.ravel())))\n",
    "\n",
    "    alleles_all_regions = {allele:sorted(set(list(itertools.chain(*alleles_all_regions[allele])))) for allele in alleles_all_regions}\n",
    "    \n",
    "    return alleles_all_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5958a2db",
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################################################################################  \n",
    "##                            Allele and Genotype counts                                   ##\n",
    "#############################################################################################\n",
    "\n",
    "\n",
    "def allele_genotype_counts(data, #Dictionary: key = region(chania,rethymno etc), value(dataframe) = allleles\n",
    "                           type_count = 'Allele', #Allele: returns allele counts, Genotype:returns observed genotype counts\n",
    "                           alleles_all_regions = None\n",
    "                          ):\n",
    "    \n",
    "    count_dict = defaultdict(lambda:defaultdict())\n",
    "    \n",
    "    if not alleles_all_regions:\n",
    "        ########################################################\n",
    "        ##   First found the unique alleles from all regions  ##\n",
    "        ########################################################\n",
    "        alleles_all_regions = unique_and_sorted_allele(data)\n",
    "\n",
    "    for region in data:\n",
    "        for allele in alleles_all_regions:\n",
    "            column_1 = allele + '_1' if 'D' in allele else allele + '1'\n",
    "            column_2 = allele + '_2' if 'D' in allele else allele + '2'\n",
    "            total_het = 0\n",
    "            total_hom = 0\n",
    "            for i in range(len(alleles_all_regions[allele])):\n",
    "\n",
    "                #Check Hom\n",
    "                count_hom = sum((data[region][column_1] == alleles_all_regions[allele][i]) & (\n",
    "                            data[region][column_2] == alleles_all_regions[allele][i]))\n",
    "                \n",
    "                if type_count == 'Allele': \n",
    "                    count_hom = 2*count_hom\n",
    "                    \n",
    "                    #Het\n",
    "                    count_het_1 = sum((data[region][column_1] == alleles_all_regions[allele][i]))\n",
    "                    count_het_2 = sum((data[region][column_2] == alleles_all_regions[allele][i]))\n",
    "                    count_het = (count_het_1 + count_het_2) - count_hom\n",
    "                else:\n",
    "                    #Het\n",
    "                    count_het = sum((data[region][column_1] == alleles_all_regions[allele][i])) - count_hom\n",
    "                    \n",
    "                total_het += count_het\n",
    "                total_hom += count_hom\n",
    "                \n",
    "                count_dict[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]] = count_hom\n",
    "                \n",
    "                #Store Het\n",
    "                count_dict[f'het_{region}_{allele}'][alleles_all_regions[allele][i]] = count_het\n",
    "                \n",
    "            count_dict[f'n_hom_{allele}_{region}'] = total_hom\n",
    "            count_dict[f'n_het_{allele}_{region}'] = total_het\n",
    "                        \n",
    "\n",
    "            #for i in range(len(alleles_all_regions[allele])):\n",
    "            #    \n",
    "            #    count_dict[f'het_rest_{region}_{allele}'][alleles_all_regions[allele][i]] = total_het - count_dict[f'het_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "            #    count_dict[f'hom_rest_{region}_{allele}'][alleles_all_regions[allele][i]] = total_hom - count_dict[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "    \n",
    "    \n",
    "    for allele in alleles_all_regions:\n",
    "        hom_total = 0\n",
    "        het_total = 0\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "            hom = 0\n",
    "            het = 0\n",
    "            for region in data:\n",
    "                hom += count_dict[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "                het += count_dict[f'het_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "                \n",
    "                 \n",
    "                count_dict[f'het_rest_{region}_{allele}'][alleles_all_regions[allele][i]] = count_dict[f'n_het_{allele}_{region}'] - count_dict[f'het_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "                count_dict[f'hom_rest_{region}_{allele}'][alleles_all_regions[allele][i]] = count_dict[f'n_hom_{allele}_{region}'] - count_dict[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]]\n",
    "\n",
    "                    \n",
    "            count_dict[f'hom_total_{allele}'][alleles_all_regions[allele][i]] = hom \n",
    "            count_dict[f'het_total_{allele}'][alleles_all_regions[allele][i]] = het\n",
    "            hom_total += hom\n",
    "            het_total += het\n",
    "            \n",
    "        count_dict[f'n_hom_total_{allele}'] = hom_total\n",
    "        count_dict[f'n_het_total_{allele}'] = het_total\n",
    "            \n",
    "\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "\n",
    "            count_dict[f'hom_rest_total_{allele}'][alleles_all_regions[allele][i]] = hom_total - count_dict[f'hom_total_{allele}'][alleles_all_regions[allele][i]]\n",
    "            count_dict[f'het_rest_total_{allele}'][alleles_all_regions[allele][i]] = het_total - count_dict[f'het_total_{allele}'][alleles_all_regions[allele][i]]\n",
    "           \n",
    "    return count_dict\n",
    "\n",
    "\n",
    "#allele_count = allele_genotype_counts(data, type_count = 'Allele')\n",
    "#obs_gen_count = allele_genotype_counts(data, type_count = 'Genotype')\n",
    "allele_count = allele_genotype_counts(data, type_count = 'Allele')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d8fed4ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##                         Step 1: Allele Frequencies                                ##\n",
    "#######################################################################################\n",
    "\n",
    "\n",
    "def help_func_freq(a,b,c,d, allele_freq, key_part1, key_part2):\n",
    "    nominator = a +b\n",
    "    denominator = c + d\n",
    "    calc_p = nominator/denominator\n",
    "    \n",
    "    allele_freq[f'p_{key_part1}'][key_part2] = calc_p\n",
    "    allele_freq[f'p^2_{key_part1}'][key_part2] = calc_p * calc_p\n",
    "    allele_freq[f'q_{key_part1}'][key_part2] = 1 - calc_p\n",
    "    allele_freq[f'q^2_{key_part1}'][key_part2] = (1 - calc_p)**2\n",
    "    \n",
    "    \n",
    "def allele_frequencies(data, #Dictionary: key (string) = region (chania,rethymno etc), value(dataframe) = allleles\n",
    "                      alleles_all_regions = None,\n",
    "                      allele_count = None\n",
    "                      ):\n",
    "    \n",
    "    n_allele = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    allele_freq = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    \n",
    "    if not alleles_all_regions and not allele_count:\n",
    "        ########################################################\n",
    "        ##   First found the unique alleles from all regions  ##\n",
    "        ########################################################\n",
    "        alleles_all_regions = unique_and_sorted_allele(data)\n",
    "\n",
    "        ########################################################\n",
    "        ##                   Allele counts                    ##\n",
    "        ########################################################\n",
    "        allele_count = allele_genotype_counts(data, type_count = 'Allele')\n",
    "\n",
    "    ########################################################\n",
    "    ##              Calculate frequencies                 ##\n",
    "    ########################################################\n",
    "\n",
    "    for allele in alleles_all_regions:\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "            total_samples = 0\n",
    "            for region in data:\n",
    "                total_samples += data[region].shape[0]\n",
    "                help_func_freq(allele_count[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                          allele_count[f'het_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                          allele_count[f'n_hom_{allele}_{region}'],\n",
    "                          allele_count[f'n_het_{allele}_{region}'], \n",
    "                          allele_freq, \n",
    "                          f\"{region}_{allele}\", \n",
    "                          alleles_all_regions[allele][i])\n",
    "                \n",
    "               \n",
    "            help_func_freq(allele_count[f'hom_total_{allele}'][alleles_all_regions[allele][i]],\n",
    "                      allele_count[f'het_total_{allele}'][alleles_all_regions[allele][i]],\n",
    "                      allele_count[f'n_hom_total_{allele}'], \n",
    "                      allele_count[f'n_het_total_{allele}'],\n",
    "                      allele_freq,\n",
    "                      f\"total_{allele}\", \n",
    "                      alleles_all_regions[allele][i])\n",
    "\n",
    "\n",
    "    return allele_freq\n",
    "\n",
    "\n",
    "#allele_freq = allele_frequencies(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "54830c4c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##             Step 2: expected genotypic counts (under HWE)                         ##\n",
    "#######################################################################################\n",
    "\n",
    "\n",
    "def help_func_hwe(a,b,c,d,counts_HWE, key_part1, key_part2):\n",
    "    counts_HWE[f'hom_p^2_{key_part1}'][key_part2] = a*b\n",
    "    counts_HWE[f'het_2pq_{key_part1}'][key_part2] = 2*c*d*b\n",
    "\n",
    "def HWE(data,\n",
    "        alleles_all_regions = None,\n",
    "        allele_freq = None\n",
    "       ):\n",
    "    \n",
    "    counts_HWE = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    \n",
    "    if not alleles_all_regions and not allele_freq:\n",
    "        ########################################################\n",
    "        ##   First found the unique alleles from all regions  ##\n",
    "        ########################################################\n",
    "        alleles_all_regions = unique_and_sorted_allele(data)\n",
    "\n",
    "        ########################################################\n",
    "        ##            Calculate allele frequencies            ##\n",
    "        ########################################################\n",
    "        allele_freq = allele_frequencies(data)\n",
    "    \n",
    "    for allele in alleles_all_regions:\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "            total_samples = 0\n",
    "            for region in data:\n",
    "                total_samples += data[region].shape[0]\n",
    "                \n",
    "                help_func_hwe(allele_freq[f'p^2_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              data[region].shape[0],\n",
    "                              allele_freq[f'p_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              allele_freq[f'q_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              counts_HWE,\n",
    "                              f'{region}_{allele}',\n",
    "                              alleles_all_regions[allele][i] \n",
    "                             )\n",
    "                \n",
    "            help_func_hwe(allele_freq[f'p^2_total_{allele}'][alleles_all_regions[allele][i]],\n",
    "                          total_samples,\n",
    "                          allele_freq[f'p_total_{allele}'][alleles_all_regions[allele][i]],\n",
    "                          allele_freq[f'q_total_{allele}'][alleles_all_regions[allele][i]],\n",
    "                          counts_HWE,\n",
    "                          f'total_{allele}',\n",
    "                          alleles_all_regions[allele][i] \n",
    "                         )\n",
    "    return counts_HWE\n",
    "              \n",
    "#counts_HWE = HWE(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85db5f75",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "86a8833f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##                 Step 3: excess or deficiencies of observed                        ##\n",
    "##      homozygotes in each subpopulation relative to HWE genotypic counts           ##\n",
    "#######################################################################################\n",
    "\n",
    "def deficiencies_sub(data,\n",
    "                     alleles_all_regions = None,\n",
    "                     obs_gen_count = None,\n",
    "                     counts_HWE = None,\n",
    "                    ):\n",
    "    \n",
    "    obs_exp = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    \n",
    "    if not alleles_all_regions and not obs_gen_count and not counts_HWE:\n",
    "        ########################################################\n",
    "        ##   First found the unique alleles from all regions  ##\n",
    "        ########################################################\n",
    "        alleles_all_regions = unique_and_sorted_allele(data)\n",
    "\n",
    "        ########################################################\n",
    "        ##                   Allele counts                    ##\n",
    "        ########################################################\n",
    "        obs_gen_count = allele_genotype_counts(data, type_count = 'Genotype')\n",
    "\n",
    "        ########################################################\n",
    "        ##                   Calculate HWE                    ##\n",
    "        ########################################################\n",
    "        counts_HWE = HWE(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##             Calculate deficiencies                 ##\n",
    "    ########################################################\n",
    "    for allele in alleles_all_regions:\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "            total_samples = 0\n",
    "            for region in data:\n",
    "    \n",
    "                obs_exp[f'obs-exp_p^2_{region}_{allele}'][alleles_all_regions[allele][i]] = (\n",
    "                                obs_gen_count[f'hom_{region}_{allele}'][alleles_all_regions[allele][i]] -\n",
    "                                counts_HWE[f'hom_p^2_{region}_{allele}'][alleles_all_regions[allele][i]])\n",
    "            obs_exp[f'obs-exp_p^2_total_{allele}'][alleles_all_regions[allele][i]] = (\n",
    "                            obs_gen_count[f'hom_total_{allele}'][alleles_all_regions[allele][i]]-\n",
    "                            counts_HWE[f'hom_p^2_total_{allele}'][alleles_all_regions[allele][i]])\n",
    "\n",
    "    return obs_exp\n",
    "\n",
    "\n",
    "#excess = deficiencies_sub(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0e8c84af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##   Step4: excess or deficiencies of observed homozygotes in the total population   ##\n",
    "#######################################################################################\n",
    "\n",
    "def deficiencies_total(data,\n",
    "                       excess_sub = None\n",
    "                      ):\n",
    "    \n",
    "    sum_obs_exp = defaultdict(lambda: 0)\n",
    "    \n",
    "    if not excess_sub:\n",
    "        ########################################################\n",
    "        ##          Calculate deficiencies_sub                ##\n",
    "        ########################################################\n",
    "        excess_sub = deficiencies_sub(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##             Calculate deficiencies                 ##\n",
    "    ########################################################\n",
    "    for key in excess_sub:\n",
    "\n",
    "        sum_obs_exp[f'sum_{key}'] = \"{:.2f}\".format(sum(excess_sub[key].values()))\n",
    "\n",
    "    return sum_obs_exp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "828fa900",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##             Step 5: local observed heterozygosities of each subpopulation         ##\n",
    "#######################################################################################\n",
    "\n",
    "\n",
    "def obs_heterozygosities(data,\n",
    "                         obs_gen_count = None,\n",
    "                        ):\n",
    "    obs_het = defaultdict(lambda:0)\n",
    "    \n",
    "    if not obs_gen_count:\n",
    "        ########################################################\n",
    "        ##                   Allele counts                    ##\n",
    "        ########################################################\n",
    "        obs_gen_count = allele_genotype_counts(data, type_count = 'Genotype')\n",
    "    \n",
    "    ########################################################\n",
    "    ##           local observed heterozygosities          ##\n",
    "    ########################################################\n",
    "    \n",
    "    for allele in data['chania'].columns[1:]:\n",
    "        total_samples = 0\n",
    "        for region in data:\n",
    "            total_samples+=data[region].shape[0]\n",
    "            allele = allele.split('_')[0] if 'D' in allele else allele[0]\n",
    "            obs_het[f'Het_{region}_{allele}'] = obs_gen_count[f'n_het_{allele}_{region}']/data[region].shape[0]\n",
    "        obs_het[f'Het_total_{allele}'] = obs_gen_count[f'n_het_total_{allele}']/total_samples\n",
    "            \n",
    "    return obs_het\n",
    "\n",
    "\n",
    "#obs_het = obs_heterozygosities(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d9c4451a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##               Step 6: expected heterozygosities of each subpopulation             ##\n",
    "#######################################################################################\n",
    "\n",
    "def exp_heterozygosities(data,\n",
    "                        allele_freq = None,\n",
    "                        ):\n",
    "    \n",
    "    exp_het = defaultdict(lambda:0)\n",
    "    \n",
    "    if not allele_freq:\n",
    "        ########################################################\n",
    "        ##            Calculate allele frequencies            ##\n",
    "        ########################################################\n",
    "        allele_freq = allele_frequencies(data)\n",
    "    \n",
    "    for allele in data['chania'].columns[1:]:\n",
    "        allele = allele.split('_')[0] if 'D' in allele else allele[0]\n",
    "        for region in data:\n",
    "            exp_het[f'Het_{region}_{allele}'] = 1 - sum(allele_freq[f'p^2_{region}_{allele}'].values())\n",
    "        exp_het[f'Het_total_{allele}'] = 1 - sum(allele_freq[f'p^2_total_{allele}'].values())\n",
    "    return exp_het\n",
    "\n",
    "\n",
    "#expected_het = exp_heterozygosities(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e469b490",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##  Step 7: local inbreeding coefficient of each subpopulation (Fs=Hexp-Hobs/Hexp)   ##\n",
    "#######################################################################################\n",
    "\n",
    "def Fs_coefficient(data,\n",
    "                   expected_het = None,\n",
    "                   obs_het = None,\n",
    "                  ):\n",
    "    \n",
    "    Fs = defaultdict(lambda: 0)\n",
    "    \n",
    "    if not expected_het and not obs_het:\n",
    "        \n",
    "        ########################################################\n",
    "        ##         Calculate expected heterozygosities        ##\n",
    "        ########################################################\n",
    "        expected_het = exp_heterozygosities(data)\n",
    "\n",
    "        ########################################################\n",
    "        ##         Calculate observed heterozygosities        ##\n",
    "        ########################################################\n",
    "        obs_het = obs_heterozygosities(data)\n",
    "    \n",
    "    for allele in data['chania'].columns[1:]:\n",
    "        allele = allele.split('_')[0] if 'D' in allele else allele[0]\n",
    "        for region in data:\n",
    "            Fs[f'Fs_{region}_{allele}'] = (expected_het[f'Het_{region}_{allele}'] - \n",
    "                                           obs_het[f'Het_{region}_{allele}'])/(\n",
    "                                           expected_het[f'Het_{region}_{allele}'])\n",
    "            \n",
    "        Fs[f'Fs_total_{allele}'] = (expected_het[f'Het_total_{allele}'] - \n",
    "                                           obs_het[f'Het_total_{allele}'])/(\n",
    "                                            expected_het[f'Het_total_{allele}'])\n",
    "            \n",
    "    return Fs\n",
    "            \n",
    "\n",
    "#fs_coef = Fs_coefficient(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b457008f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##            Step8: p_allele bar (global frequency of each allele                   ##\n",
    "##       across all subpopulations and weighted by each subpopulation size)          ##\n",
    "#######################################################################################\n",
    "\n",
    "\n",
    "def help_func_bar(a,b,c,d,allele_bar, key_part1, key_part2):\n",
    "    \n",
    "    if d == 0:\n",
    "        denominator = b\n",
    "        nominator = c*a\n",
    "        allele_bar[f'{key_part1}'][key_part2] += nominator/denominator\n",
    "        allele_bar[f'square_{key_part1}'][key_part2] = allele_bar[f'{key_part1}'][key_part2]**2\n",
    "    else: \n",
    "        denominator = a+b\n",
    "        nominator = c*a + d*b\n",
    "        allele_bar[f'{key_part1}'][key_part2] = nominator/denominator\n",
    "        allele_bar[f'square_{key_part1}'][key_part2] = (nominator/denominator)**2\n",
    "\n",
    "def p_allele_bar(data,\n",
    "                 alleles_all_regions = None,\n",
    "                 allele_freq = None,\n",
    "                 allele_count = None\n",
    "                ):\n",
    "    \n",
    "    n_allele = defaultdict(lambda: 0)\n",
    "    allele_bar = defaultdict(lambda: defaultdict(lambda: 0))\n",
    "    \n",
    "    if not alleles_all_regions and not allele_freq and not allele_count:\n",
    "    \n",
    "        ########################################################\n",
    "        ##   First found the unique alleles from all regions  ##\n",
    "        ########################################################\n",
    "        alleles_all_regions = unique_and_sorted_allele(data)\n",
    "\n",
    "        ########################################################\n",
    "        ##            Calculate allele frequencies            ##\n",
    "        ########################################################\n",
    "        allele_freq = allele_frequencies(data)\n",
    "\n",
    "        ########################################################\n",
    "        ##                   Allele counts                    ##\n",
    "        ########################################################\n",
    "        allele_count = allele_genotype_counts(data, type_count = 'Allele')\n",
    "    \n",
    "    total_samples = sum([data[region].shape[0] for region in data])*2\n",
    "    for allele in alleles_all_regions:\n",
    "        for region in data:\n",
    "            for i in range(len(alleles_all_regions[allele])):\n",
    "                \n",
    "                help_func_bar(data[region].shape[0]*2,\n",
    "                              total_samples,\n",
    "                              allele_freq[f'p_{region}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              0,\n",
    "                              allele_bar,\n",
    "                              f'all_{allele}',\n",
    "                              alleles_all_regions[allele][i]\n",
    "                             )\n",
    "\n",
    "\n",
    "        for i in range(len(alleles_all_regions[allele])):\n",
    "            for region_1, region_2 in list(itertools.combinations(data.keys(), 2)):\n",
    "\n",
    "                help_func_bar(data[region_1].shape[0]*2,\n",
    "                              data[region_2].shape[0]*2,\n",
    "                              allele_freq[f'p_{region_1}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              allele_freq[f'p_{region_2}_{allele}'][alleles_all_regions[allele][i]],\n",
    "                              allele_bar,\n",
    "                              f'{region_1}_{region_2}_{allele}',\n",
    "                              alleles_all_regions[allele][i]\n",
    "                             )\n",
    "\n",
    "    return allele_bar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "36ff19ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##                   Step9&10&11: heterozygosity indices                             ##\n",
    "##              (Hi=weighted average of observed heterozygosities),                  ##\n",
    "##              (Hs =weighted average of expected heterozygosities),                 ##\n",
    "## (Ht=expected hetrozygosty based on the global gene frequencies i.e. p_allele-bar) ##\n",
    "#######################################################################################\n",
    "\n",
    "def heterozygosity_indices(data,\n",
    "                           expected_het = None,\n",
    "                           obs_het = None,\n",
    "                           allele_bar = None,\n",
    "                          ):\n",
    "    \n",
    "    Het_indices = defaultdict(lambda: 0)\n",
    "    \n",
    "    if not expected_het and not obs_het and not allele_bar:\n",
    "        ########################################################\n",
    "        ##         Calculate expected heterozygosities        ##\n",
    "        ########################################################\n",
    "        expected_het = exp_heterozygosities(data)\n",
    "\n",
    "        ########################################################\n",
    "        ##         Calculate observed heterozygosities        ##\n",
    "        ########################################################\n",
    "        obs_het = obs_heterozygosities(data)\n",
    "\n",
    "        ########################################################\n",
    "        ##         Calculate observed heterozygosities        ##\n",
    "        ########################################################\n",
    "        allele_bar = p_allele_bar(data)\n",
    "\n",
    "    \n",
    "    for allele in data['chania'].columns[1:]:\n",
    "        total_samples = 0\n",
    "        allele = allele.split('_')[0] if 'D' in allele else allele[0]\n",
    "        for region in data:\n",
    "            total_samples+=data[region].shape[0]\n",
    "            Het_indices[f'Hi_all_{allele}'] += (obs_het[f'Het_{region}_{allele}'] * data[region].shape[0])\n",
    "            Het_indices[f'Hs_all_{allele}'] += (expected_het[f'Het_{region}_{allele}'] * data[region].shape[0])\n",
    "            \n",
    "        Het_indices[f'Hi_all_{allele}'] /= total_samples\n",
    "        Het_indices[f'Hs_all_{allele}'] /= total_samples\n",
    "        Het_indices[f'Ht_all_{allele}'] = 1 - sum(allele_bar[f'square_all_{allele}'].values())\n",
    "        \n",
    "\n",
    "        for region_1, region_2 in list(itertools.combinations(data.keys(), 2)):  \n",
    "            Het_indices[f'Hi_{region_1}_{region_2}_{allele}'] = (\n",
    "                                                    (obs_het[f'Het_{region_1}_{allele}'] * data[region_1].shape[0])\n",
    "                                                     + (obs_het[f'Het_{region_2}_{allele}'] * data[region_2].shape[0]))\\\n",
    "                                                     / (data[region_1].shape[0] + data[region_2].shape[0])\n",
    "\n",
    "            Het_indices[f'Hs_{region_1}_{region_2}_{allele}'] = (\n",
    "                                                    (expected_het[f'Het_{region_1}_{allele}'] * data[region_1].shape[0])\n",
    "                                                 + (expected_het[f'Het_{region_2}_{allele}'] * data[region_2].shape[0]))\\\n",
    "                                                 / (data[region_1].shape[0] + data[region_2].shape[0])\n",
    "            Het_indices[f'Ht_{region_1}_{region_2}_{allele}'] = 1 - \\\n",
    "                                    sum(allele_bar[f'square_{region_1}_{region_2}_{allele}'].values())\n",
    "            \n",
    "    return Het_indices\n",
    " \n",
    "    \n",
    "    \n",
    "#Het_indices = heterozygosity_indices(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3c55700a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################################\n",
    "##           Step12: global Fstatistics (FIS_observed heterozygosities,              ## \n",
    "##                    FST_expected heterozygosities,                                 ##\n",
    "##                    FIT_expected heterozygosities)                                 ##\n",
    "#######################################################################################\n",
    "\n",
    "def help_func_Fstats(a,b,c, Fstatistics, key_part):\n",
    "    \n",
    "    Fstatistics[f'Fis_{key_part}'] = (a-b)/a\n",
    "    Fstatistics[f'Fst_{key_part}'] = (c-a)/c\n",
    "    Fstatistics[f'Fit_{key_part}'] = (c-b)/c\n",
    "\n",
    "def global_Fstatistics(data,\n",
    "                       Het_indices = None,\n",
    "                      ):\n",
    "    \n",
    "    Fstatistics = defaultdict(lambda:0)\n",
    "    \n",
    "    if not Het_indices:\n",
    "        ########################################################\n",
    "        ##          Calculate heterozygosity indices          ##\n",
    "        ########################################################\n",
    "        Het_indices = heterozygosity_indices(data)\n",
    "    \n",
    "    for allele in data['chania'].columns[1:]:\n",
    "        allele = allele.split('_')[0] if 'D' in allele else allele[0]\n",
    "        help_func_Fstats(Het_indices[f'Hs_all_{allele}'],\n",
    "                         Het_indices[f'Hi_all_{allele}'],\n",
    "                         Het_indices[f'Ht_all_{allele}'],\n",
    "                         Fstatistics,\n",
    "                         f'all_{allele}',\n",
    "                         )\n",
    "        \n",
    "        for region_1, region_2 in list(itertools.combinations(data.keys(), 2)):\n",
    "            \n",
    "            help_func_Fstats(Het_indices[f'Hs_{region_1}_{region_2}_{allele}'],\n",
    "                             Het_indices[f'Hi_{region_1}_{region_2}_{allele}'],\n",
    "                             Het_indices[f'Ht_{region_1}_{region_2}_{allele}'],\n",
    "                             Fstatistics,\n",
    "                             f'_{region_1}_{region_2}_{allele}',\n",
    "                            )\n",
    "    return Fstatistics\n",
    "    \n",
    "        \n",
    "#Fstatistics = global_Fstatistics(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b7b6ea2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fst_pipeline(data):\n",
    "    ########################################################\n",
    "    ##   First found the unique alleles from all regions  ##\n",
    "    ########################################################\n",
    "    alleles_all_regions = unique_and_sorted_allele(data)\n",
    "    \n",
    "    ########################################################\n",
    "    ##            Allele and Genotype counts              ##\n",
    "    ########################################################\n",
    "    allele_count = allele_genotype_counts(data, \n",
    "                                          'Allele', \n",
    "                                          alleles_all_regions)\n",
    "    \n",
    "    obs_gen_count = allele_genotype_counts(data, \n",
    "                                          'Genotype', \n",
    "                                           alleles_all_regions\n",
    "                                          )\n",
    "    \n",
    "    ########################################################\n",
    "    ##            Calculate allele frequencies            ##\n",
    "    ########################################################\n",
    "    allele_freq = allele_frequencies(data,\n",
    "                                     alleles_all_regions,\n",
    "                                     allele_count\n",
    "                                    )\n",
    "    \n",
    "    ########################################################\n",
    "    ##                   Calculate HWE                    ##\n",
    "    ########################################################\n",
    "    counts_HWE = HWE(data,\n",
    "                     alleles_all_regions,\n",
    "                     allele_freq\n",
    "                    )\n",
    "    \n",
    "    ########################################################\n",
    "    ##          Calculate deficiencies_sub                ##\n",
    "    ########################################################\n",
    "    excess_sub = deficiencies_sub(data,\n",
    "                                  alleles_all_regions,\n",
    "                                  obs_gen_count,\n",
    "                                  counts_HWE\n",
    "                                 )\n",
    "    \n",
    "    ########################################################\n",
    "    ##          Calculate deficiencies_sub                ##\n",
    "    ########################################################\n",
    "    excess_total = deficiencies_total(data,\n",
    "                                      excess_sub\n",
    "                                     )\n",
    "    \n",
    "    ########################################################\n",
    "    ##         Calculate observed heterozygosities        ##\n",
    "    ########################################################\n",
    "    obs_het = obs_heterozygosities(data,\n",
    "                                   obs_gen_count\n",
    "                                  )\n",
    " \n",
    "    ########################################################\n",
    "    ##         Calculate expected heterozygosities        ##\n",
    "    ########################################################\n",
    "    expected_het = exp_heterozygosities(data,\n",
    "                                        allele_freq\n",
    "                                       )\n",
    "    \n",
    "    ########################################################\n",
    "    ##               Calculate allele bar                 ##\n",
    "    ########################################################\n",
    "    allele_bar = p_allele_bar(data,\n",
    "                              alleles_all_regions,\n",
    "                              allele_freq,\n",
    "                              allele_count\n",
    "                             )\n",
    "    \n",
    "    ########################################################\n",
    "    ##          Calculate heterozygosity indices          ##\n",
    "    ########################################################\n",
    "    Het_indices = heterozygosity_indices(data,\n",
    "                                         expected_het,\n",
    "                                         obs_het,\n",
    "                                         allele_bar\n",
    "                                        )\n",
    "\n",
    "    ########################################################\n",
    "    ##            Calculate global Fstatistic             ##\n",
    "    ########################################################\n",
    "\n",
    "    Fstatistics = global_Fstatistics(data,\n",
    "                                     Het_indices\n",
    "                                    )\n",
    "    \n",
    "    return (alleles_all_regions, allele_count, \n",
    "            obs_gen_count, allele_freq,counts_HWE,\n",
    "            excess_sub,excess_total,obs_het,expected_het,\n",
    "            allele_bar, Het_indices, Fstatistics)\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1941590e",
   "metadata": {},
   "outputs": [],
   "source": [
    "(alleles_all_regions, allele_count, \n",
    "obs_gen_count, allele_freq,counts_HWE,\n",
    "excess_sub,excess_total,obs_het,expected_het,\n",
    "allele_bar, Het_indices, Fstatistics) = fst_pipeline(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b3d4e99",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c8aa832",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "599c0cf0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3eebd0e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################\n",
    "## store allele_counts and gen_counts ##\n",
    "########################################\n",
    "\n",
    "results_dataframe = pd.concat([pd.DataFrame([ f\"{key}*{element}\"for element in alleles_all_regions[key]]) for key in alleles_all_regions.keys()])\n",
    "for key in set([\"_\".join(values.split(\"_\")[:-1]) for values in allele_count if not values.startswith(\"n_\")]):\n",
    "    results_dataframe['allele_' + key] = [allele_count[key + \"_\" + item.split(\"*\")[0]][item.split(\"*\")[1]] for item in list(results_dataframe.iloc[:,0].values)]\n",
    "    results_dataframe['gen_' + key] = [obs_gen_count[key + \"_\" + item.split(\"*\")[0]][item.split(\"*\")[1]] for item in list(results_dataframe.iloc[:,0].values)]\n",
    "\n",
    "for key in set([\"_\".join(values.split(\"_\")[:-1]) for values in allele_freq]):\n",
    "    results_dataframe[key] = [allele_freq[key + \"_\" + item.split(\"*\")[0]][item.split(\"*\")[1]] for item in list(results_dataframe.iloc[:,0].values)]\n",
    "    \n",
    "for key in set([\"_\".join(values.split(\"_\")[:-1]) for values in counts_HWE]):\n",
    "    results_dataframe['HWE_'+key] = [counts_HWE[key + \"_\" + item.split(\"*\")[0]][item.split(\"*\")[1]] for item in list(results_dataframe.iloc[:,0].values)]\n",
    "    \n",
    "for key in set([\"_\".join(values.split(\"_\")[:-1]) for values in excess_sub]):\n",
    "    results_dataframe[key] = [excess_sub[key + \"_\" + item.split(\"*\")[0]][item.split(\"*\")[1]] for item in list(results_dataframe.iloc[:,0].values)]\n",
    "\n",
    "for key in set([\"_\".join(values.split(\"_\")[:-1]) for values in allele_bar]):\n",
    "    results_dataframe['allele_bar_'+key] = [allele_bar[key + \"_\" + item.split(\"*\")[0]][item.split(\"*\")[1]] for item in list(results_dataframe.iloc[:,0].values)]\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ba5b495a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>allele_hom_total</th>\n",
       "      <th>gen_hom_total</th>\n",
       "      <th>allele_het_rest_total</th>\n",
       "      <th>gen_het_rest_total</th>\n",
       "      <th>allele_het_lasithi</th>\n",
       "      <th>gen_het_lasithi</th>\n",
       "      <th>allele_hom_rest_lasithi</th>\n",
       "      <th>gen_hom_rest_lasithi</th>\n",
       "      <th>allele_hom_rest_heraklion</th>\n",
       "      <th>...</th>\n",
       "      <th>allele_bar_square_chania_rethymno</th>\n",
       "      <th>allele_bar_rethymno_heraklion</th>\n",
       "      <th>allele_bar_chania_heraklion</th>\n",
       "      <th>allele_bar_rethymno_lasithi</th>\n",
       "      <th>allele_bar_heraklion_lasithi</th>\n",
       "      <th>allele_bar_square_chania_lasithi</th>\n",
       "      <th>allele_bar_chania_rethymno</th>\n",
       "      <th>allele_bar_square_rethymno_heraklion</th>\n",
       "      <th>allele_bar_square_chania_heraklion</th>\n",
       "      <th>allele_bar_chania_lasithi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A*01:01</td>\n",
       "      <td>48</td>\n",
       "      <td>24</td>\n",
       "      <td>2812</td>\n",
       "      <td>1249</td>\n",
       "      <td>56</td>\n",
       "      <td>56</td>\n",
       "      <td>74</td>\n",
       "      <td>37</td>\n",
       "      <td>96</td>\n",
       "      <td>...</td>\n",
       "      <td>9.200333e-03</td>\n",
       "      <td>0.114301</td>\n",
       "      <td>0.107679</td>\n",
       "      <td>0.098200</td>\n",
       "      <td>0.110505</td>\n",
       "      <td>8.508067e-03</td>\n",
       "      <td>0.095918</td>\n",
       "      <td>1.306463e-02</td>\n",
       "      <td>1.159471e-02</td>\n",
       "      <td>0.092239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A*01:02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3125</td>\n",
       "      <td>1561</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>40</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>3.188776e-06</td>\n",
       "      <td>0.001044</td>\n",
       "      <td>0.001324</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003021</td>\n",
       "      <td>2.281834e-06</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>1.089605e-06</td>\n",
       "      <td>1.752761e-06</td>\n",
       "      <td>0.001511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A*01:03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3120</td>\n",
       "      <td>1558</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>80</td>\n",
       "      <td>40</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>4.164931e-06</td>\n",
       "      <td>0.001044</td>\n",
       "      <td>0.001324</td>\n",
       "      <td>0.004092</td>\n",
       "      <td>0.002478</td>\n",
       "      <td>1.456792e-05</td>\n",
       "      <td>0.002041</td>\n",
       "      <td>1.089605e-06</td>\n",
       "      <td>1.752761e-06</td>\n",
       "      <td>0.003817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A*02:01</td>\n",
       "      <td>144</td>\n",
       "      <td>72</td>\n",
       "      <td>2598</td>\n",
       "      <td>1107</td>\n",
       "      <td>109</td>\n",
       "      <td>95</td>\n",
       "      <td>44</td>\n",
       "      <td>22</td>\n",
       "      <td>70</td>\n",
       "      <td>...</td>\n",
       "      <td>3.225323e-02</td>\n",
       "      <td>0.180585</td>\n",
       "      <td>0.197705</td>\n",
       "      <td>0.184943</td>\n",
       "      <td>0.203171</td>\n",
       "      <td>4.353541e-02</td>\n",
       "      <td>0.179592</td>\n",
       "      <td>3.261078e-02</td>\n",
       "      <td>3.908735e-02</td>\n",
       "      <td>0.208651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A*02:02</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3104</td>\n",
       "      <td>1545</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>40</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>1.184692e-04</td>\n",
       "      <td>0.004175</td>\n",
       "      <td>0.009709</td>\n",
       "      <td>0.001637</td>\n",
       "      <td>0.003964</td>\n",
       "      <td>1.035941e-04</td>\n",
       "      <td>0.010884</td>\n",
       "      <td>1.743368e-05</td>\n",
       "      <td>9.425959e-05</td>\n",
       "      <td>0.010178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>DPB1*701:01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2806</td>\n",
       "      <td>1404</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>56</td>\n",
       "      <td>290</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000522</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.000818</td>\n",
       "      <td>0.000991</td>\n",
       "      <td>4.046643e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.724012e-07</td>\n",
       "      <td>1.947512e-07</td>\n",
       "      <td>0.000636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>DPB1*791:01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2807</td>\n",
       "      <td>1404</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>56</td>\n",
       "      <td>290</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000818</td>\n",
       "      <td>0.000496</td>\n",
       "      <td>4.046643e-07</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>DPB1*80:01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2807</td>\n",
       "      <td>1404</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>56</td>\n",
       "      <td>290</td>\n",
       "      <td>...</td>\n",
       "      <td>3.188776e-06</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000737</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.281834e-06</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.438519e-07</td>\n",
       "      <td>0.001511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>DPB1*81:01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2807</td>\n",
       "      <td>1404</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>56</td>\n",
       "      <td>290</td>\n",
       "      <td>...</td>\n",
       "      <td>3.188776e-06</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000737</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.281834e-06</td>\n",
       "      <td>0.001786</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.438519e-07</td>\n",
       "      <td>0.001511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>DPB1*882:01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2806</td>\n",
       "      <td>1403</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>56</td>\n",
       "      <td>290</td>\n",
       "      <td>...</td>\n",
       "      <td>4.627701e-07</td>\n",
       "      <td>0.001044</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.001511</td>\n",
       "      <td>0.001511</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000680</td>\n",
       "      <td>1.089605e-06</td>\n",
       "      <td>1.947512e-07</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246 rows  90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0  allele_hom_total  gen_hom_total  allele_het_rest_total  \\\n",
       "0       A*01:01                48             24                   2812   \n",
       "1       A*01:02                 0              0                   3125   \n",
       "2       A*01:03                 0              0                   3120   \n",
       "3       A*02:01               144             72                   2598   \n",
       "4       A*02:02                 0              0                   3104   \n",
       "..          ...               ...            ...                    ...   \n",
       "36  DPB1*701:01                 0              0                   2806   \n",
       "37  DPB1*791:01                 0              0                   2807   \n",
       "38   DPB1*80:01                 0              0                   2807   \n",
       "39   DPB1*81:01                 0              0                   2807   \n",
       "40  DPB1*882:01                 0              0                   2806   \n",
       "\n",
       "    gen_het_rest_total  allele_het_lasithi  gen_het_lasithi  \\\n",
       "0                 1249                  56               56   \n",
       "1                 1561                   0                0   \n",
       "2                 1558                   4                3   \n",
       "3                 1107                 109               95   \n",
       "4                 1545                   1                1   \n",
       "..                 ...                 ...              ...   \n",
       "36                1404                   1                0   \n",
       "37                1404                   1                0   \n",
       "38                1404                   0                0   \n",
       "39                1404                   0                0   \n",
       "40                1403                   0                0   \n",
       "\n",
       "    allele_hom_rest_lasithi  gen_hom_rest_lasithi  allele_hom_rest_heraklion  \\\n",
       "0                        74                    37                         96   \n",
       "1                        80                    40                        118   \n",
       "2                        80                    40                        118   \n",
       "3                        44                    22                         70   \n",
       "4                        80                    40                        118   \n",
       "..                      ...                   ...                        ...   \n",
       "36                      112                    56                        290   \n",
       "37                      112                    56                        290   \n",
       "38                      112                    56                        290   \n",
       "39                      112                    56                        290   \n",
       "40                      112                    56                        290   \n",
       "\n",
       "    ...  allele_bar_square_chania_rethymno  allele_bar_rethymno_heraklion  \\\n",
       "0   ...                       9.200333e-03                       0.114301   \n",
       "1   ...                       3.188776e-06                       0.001044   \n",
       "2   ...                       4.164931e-06                       0.001044   \n",
       "3   ...                       3.225323e-02                       0.180585   \n",
       "4   ...                       1.184692e-04                       0.004175   \n",
       "..  ...                                ...                            ...   \n",
       "36  ...                       0.000000e+00                       0.000522   \n",
       "37  ...                       0.000000e+00                       0.000000   \n",
       "38  ...                       3.188776e-06                       0.000000   \n",
       "39  ...                       3.188776e-06                       0.000000   \n",
       "40  ...                       4.627701e-07                       0.001044   \n",
       "\n",
       "    allele_bar_chania_heraklion  allele_bar_rethymno_lasithi  \\\n",
       "0                      0.107679                     0.098200   \n",
       "1                      0.001324                     0.000000   \n",
       "2                      0.001324                     0.004092   \n",
       "3                      0.197705                     0.184943   \n",
       "4                      0.009709                     0.001637   \n",
       "..                          ...                          ...   \n",
       "36                     0.000441                     0.000818   \n",
       "37                     0.000000                     0.000818   \n",
       "38                     0.000737                     0.000000   \n",
       "39                     0.000737                     0.000000   \n",
       "40                     0.000441                     0.001511   \n",
       "\n",
       "    allele_bar_heraklion_lasithi  allele_bar_square_chania_lasithi  \\\n",
       "0                       0.110505                      8.508067e-03   \n",
       "1                       0.003021                      2.281834e-06   \n",
       "2                       0.002478                      1.456792e-05   \n",
       "3                       0.203171                      4.353541e-02   \n",
       "4                       0.003964                      1.035941e-04   \n",
       "..                           ...                               ...   \n",
       "36                      0.000991                      4.046643e-07   \n",
       "37                      0.000496                      4.046643e-07   \n",
       "38                      0.000000                      2.281834e-06   \n",
       "39                      0.000000                      2.281834e-06   \n",
       "40                      0.001511                      0.000000e+00   \n",
       "\n",
       "    allele_bar_chania_rethymno  allele_bar_square_rethymno_heraklion  \\\n",
       "0                     0.095918                          1.306463e-02   \n",
       "1                     0.001786                          1.089605e-06   \n",
       "2                     0.002041                          1.089605e-06   \n",
       "3                     0.179592                          3.261078e-02   \n",
       "4                     0.010884                          1.743368e-05   \n",
       "..                         ...                                   ...   \n",
       "36                    0.000000                          2.724012e-07   \n",
       "37                    0.000000                          0.000000e+00   \n",
       "38                    0.001786                          0.000000e+00   \n",
       "39                    0.001786                          0.000000e+00   \n",
       "40                    0.000680                          1.089605e-06   \n",
       "\n",
       "    allele_bar_square_chania_heraklion  allele_bar_chania_lasithi  \n",
       "0                         1.159471e-02                   0.092239  \n",
       "1                         1.752761e-06                   0.001511  \n",
       "2                         1.752761e-06                   0.003817  \n",
       "3                         3.908735e-02                   0.208651  \n",
       "4                         9.425959e-05                   0.010178  \n",
       "..                                 ...                        ...  \n",
       "36                        1.947512e-07                   0.000636  \n",
       "37                        0.000000e+00                   0.000636  \n",
       "38                        5.438519e-07                   0.001511  \n",
       "39                        5.438519e-07                   0.001511  \n",
       "40                        1.947512e-07                   0.000000  \n",
       "\n",
       "[246 rows x 90 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2163cdb1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
